<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Understand outputs · RABBIT</title><meta name="title" content="Understand outputs · RABBIT"/><meta property="og:title" content="Understand outputs · RABBIT"/><meta property="twitter:title" content="Understand outputs · RABBIT"/><meta name="description" content="Documentation for RABBIT."/><meta property="og:description" content="Documentation for RABBIT."/><meta property="twitter:description" content="Documentation for RABBIT."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">RABBIT</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2"><span class="docs-label">Guide</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../run_rabbit/">Run RABBIT</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Prepare Inputs</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../prepare_input/prepare_genofile/">Prepare genofile</a></li><li><a class="tocitem" href="../prepare_input/prepare_pedinfo_designcode/">Prepare pedinfo: designcode</a></li><li><a class="tocitem" href="../prepare_input/prepare_pedinfo_pedfile/">Prepare pedinfo: pedfile</a></li><li><a class="tocitem" href="../prepare_input/prepare_phenofile/">Prepare phenofile</a></li><li><a class="tocitem" href="../prepare_input/prepare_convert/">Prepare by converting </a></li></ul></li><li class="is-active"><a class="tocitem" href>Understand outputs</a><ul class="internal"><li><a class="tocitem" href="#Overview"><span>Overview</span></a></li><li><a class="tocitem" href="#MagicMap"><span>MagicMap</span></a></li><li><a class="tocitem" href="#MagicReconstruct"><span>MagicReconstruct</span></a></li></ul></li><li><a class="tocitem" href="../key_options/">Key options</a></li><li><a class="tocitem" href="../statistical_framework/">Statistical framework</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox"/><label class="tocitem" for="menuitem-3"><span class="docs-label">Pipeline</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../pipeline/s0_pipeline_overview/">Overview</a></li><li><a class="tocitem" href="../pipeline/s1_magicsimulate/">S1 MagicSimulate</a></li><li><a class="tocitem" href="../pipeline/s2_magicfilter/">S2 MagicFilter</a></li><li><a class="tocitem" href="../pipeline/s3_magiccall/">S3 MagicCall</a></li><li><a class="tocitem" href="../pipeline/s4_magicmap/">S4 MagicMap</a></li><li><a class="tocitem" href="../pipeline/s5_magicimpute/">S5 MagicImpute</a></li><li><a class="tocitem" href="../pipeline/s6_magicreconstruct/">S6 MagicReconstruct</a></li><li><a class="tocitem" href="../pipeline/s7_magicscan/">S7 MagicScan</a></li></ul></li><li><a class="tocitem" href="../rabbit_cli/">CLI</a></li><li><a class="tocitem" href="../rabbit_api/">API</a></li><li><a class="tocitem" href="../ref/">Citations</a></li></ul><div class="docs-version-selector field has-addons visible"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"><option value="#" selected="selected">v1.0.0</option></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Guide</a></li><li class="is-active"><a href>Understand outputs</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Understand outputs</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Understand-output"><a class="docs-heading-anchor" href="#Understand-output">Understand output</a><a id="Understand-output-1"></a><a class="docs-heading-anchor-permalink" href="#Understand-output" title="Permalink"></a></h1><h2 id="Overview"><a class="docs-heading-anchor" href="#Overview">Overview</a><a id="Overview-1"></a><a class="docs-heading-anchor-permalink" href="#Overview" title="Permalink"></a></h2><p><img src="../RABBIT_pipeline.png" alt/></p><div class="admonition is-info"><header class="admonition-header">Naming output files</header><div class="admonition-body"><ul><li>Each output filename can be split into: outstem, function name, individual file ID, and file extension, which are joined by underscores &#39;_&#39;. The first part outstem is specified via keyword <code>outstem</code>. </li><li><code>logfile</code>: <code>outstem*&quot;_&quot;*funcname*&quot;.log&quot;</code></li><li><code>genofile</code>: <code>outstem*&quot;_&quot;*funcname*&quot;_geno.vcf.gz&quot;</code></li><li><code>pedfile</code>: <code>outstem*&quot;_&quot;*funcname*&quot;_ped.csv&quot;</code></li></ul></div></div><p>The RABBIT output files can be categorized into four groups: </p><ul><li><code>logfile</code>.   Information include input argument values, output filenames, and computational time. </li><li><code>genofile</code>  and <code>pedfile</code>.  See the above flowchart and the section <a href="../prepare_input/prepare_genofile/#prepare-genofile"><code>Prepare Inputs</code></a>. </li><li>PNG plot files. See section <a href="../pipeline/s0_pipeline_overview/#overview"><code>Pipeline</code></a>  </li><li>CSV text files (except <code>pedfile</code>). </li></ul><p>The main output files of MagicSimulate, MagicFilter, MagicCall, and MagicImpute are <code>genofile</code> and/or <code>pedfile</code>, which have the same formats as the input files. Most CSV output files contain a single table, each column being explained in the beginning comment lines (starting with &quot;##&quot;).</p><p>In the following, we will mainly describe the three CSV text files that contain multiple tables: <code>ldfile</code> and <code>linkagefile</code> resulting from MagicMap and <code>ancestryfile</code> resulting from MagicReconstruct. In each of these files, each table starts with a row including only two cells: &quot;RABBIT&quot; and table name, followed by a header row and then content rows. </p><h2 id="MagicMap"><a class="docs-heading-anchor" href="#MagicMap">MagicMap</a><a id="MagicMap-1"></a><a class="docs-heading-anchor-permalink" href="#MagicMap" title="Permalink"></a></h2><h3 id="ldfile"><a class="docs-heading-anchor" href="#ldfile">ldfile</a><a id="ldfile-1"></a><a class="docs-heading-anchor-permalink" href="#ldfile" title="Permalink"></a></h3><p><code>MagicMap</code> performs pairwise linkage disequilibrium (LD) analyses via <code>MagicLD.magicld</code>. If marker binning is performed (<code>isbinning = true</code>), LD analyses are only for the representative markers from each bin. The output <code>ldfile</code> (<code>outstem*&quot;_magicmap_magicld.csv.gz</code>) contains three tables: </p><ul><li><code>offspringinfo</code>: same as the <a href="../prepare_input/prepare_pedinfo_pedfile/#offspringinfo"><code>offspringinfo</code></a> of <code>pedfile</code></li><li><code>markerinfo</code>: table with each row denoting a (representative) marker. It consists of 5 columns:<ul><li><code>markerno</code>: row index starting from 1</li><li><code>marker_represent</code>: representative marker ID </li><li><code>physchrom_represent</code>: physical map chromosome ID for the representative marker </li><li><code>physposbp_represent</code>: physical map position (in base pair) for the representative marker</li><li><code>ld_bin</code>:  list of marker IDs in the bin containing the representative marker. &quot;NA&quot; if <code>isbinning = false</code></li></ul></li><li><code>pairwiseld</code>: table with each row denoting LD analysis for two representative markers. It consists of 4 columns:<ul><li><code>marker1</code>: index of 1st representative marker, corresponding to <code>markerno</code> in table <code>markerinfo</code></li><li><code>marker2</code>: index of 2nd representative marker, corresponding to <code>markerno</code> in table <code>markerinfo</code></li><li><code>ld_r2</code>: squared allelic correlation between the two representative markers</li><li><code>ld_lod</code>: LOD score of the LD analysis for the two representative markers</li></ul></li></ul><p>The results are saved only if <code>ld_r2 &gt;= minldsave</code> and <code>ld_lod &gt;= minlodsave</code>. By default, the values of keyargs <code>minldsave</code> and <code>minlodsave</code> are nothing and they are reset internally  by <code>MagicLD.magicld</code> according to the number of (representative) markers; see the logfile for the reset values. </p><h3 id="linkagefile"><a class="docs-heading-anchor" href="#linkagefile">linkagefile</a><a id="linkagefile-1"></a><a class="docs-heading-anchor-permalink" href="#linkagefile" title="Permalink"></a></h3><p><code>MagicMap</code> performs pairwise linkage analyses via <code>MagicLinkage.magiclinkage</code> for the pairs of (representative) marker in the <code>ldfile</code>. The output <code>linkagefile</code> (<code>outstem*&quot;_magicmap_magiclinkage.csv.gz</code>) contains three tables: </p><ul><li><code>offspringinfo</code>: same as the <a href="../prepare_input/prepare_pedinfo_pedfile/#offspringinfo"><code>offspringinfo</code></a> of <code>pedfile</code></li><li><code>markerinfo</code>: table with each row denoting a (representative) marker. It consists of 5 columns:<ul><li><code>markerno</code>: row index starting from 1</li><li><code>marker</code>: marker ID</li><li><code>physchrom</code>: physical map chromosome ID for the marker </li><li><code>physposbp</code>: physical map position (in base pair) for the marker    </li><li><code>nmissing</code>:  number of missing genotypes at the marker</li></ul></li><li><code>pairwiselinkage</code>: table with each row denoting linkage analysis for two markers. It consists of 4 columns:<ul><li><code>marker1</code>: index of 1st marker, corresponding to <code>markerno</code> in table <code>markerinfo</code></li><li><code>marker2</code>: index of 2nd marker, corresponding to <code>markerno</code> in table <code>markerinfo</code></li><li><code>linkage_rf</code>: recombination fraction between the two markers (scaled from 0 to 1)</li><li><code>linkage_lod</code>: LOD score of the linkage analysis for the two  markers</li></ul></li></ul><p>The results are saved only if <code>linkage_rf &gt;= maxrfsave</code> and <code>linkage_lod &gt;= minlodsave</code>. By default, the value of keyarg <code>maxrfsave</code> and <code>minlodsave</code> are nothing and they are reset internally by <code>MagicLinkage.magiclinkage</code> according to the number of (representative) markers; see the logfile for the reset values. </p><h2 id="MagicReconstruct"><a class="docs-heading-anchor" href="#MagicReconstruct">MagicReconstruct</a><a id="MagicReconstruct-1"></a><a class="docs-heading-anchor-permalink" href="#MagicReconstruct" title="Permalink"></a></h2><h3 id="ancestryfile"><a class="docs-heading-anchor" href="#ancestryfile">ancestryfile</a><a id="ancestryfile-1"></a><a class="docs-heading-anchor-permalink" href="#ancestryfile" title="Permalink"></a></h3><p>The output <code>ancestryfile</code> (<code>outstem*&quot;_magicreconstruct_ancestry.csv.gz&quot;</code>) is a CSV file containing 13 tables:</p><ul><li><code>designinfo</code>: same as the <a href="../prepare_input/prepare_pedinfo_pedfile/#designinfo"><code>designinfo</code></a> of <code>pedfile</code> </li><li><code>founderinfo</code>: founder information. <ul><li><code>individual</code>: founder ID</li><li><code>gender</code>: &quot;notapplicable&quot;</li></ul></li><li><code>offspringinfo</code>: same as the <a href="../prepare_input/prepare_pedinfo_pedfile/#offspringinfo"><code>offspringinfo</code></a> of <code>pedfile</code></li><li><code>foundergeno</code>: genotypic data for founders. Columns 1-13 are the same as those in <code>refinedmapfile</code>, and each of the rest columns is genotypic data for each founder.</li><li><code>haplotype</code>: ancestral haplotype states<ul><li><code>haplotypeindex</code>: row index starting from 1</li><li><code>haploindex</code>: haplotype index, same as <code>haplotypeindex</code></li><li><code>haplotype</code>: founder genomic labels. If a founder is inbred, its label is the founderID. If a founder is outbred, it has two labels: founderID<code>_1</code> and founderID<code>_2</code> for each of the two gametes. </li></ul></li><li><code>genotype</code>: ancestral genotype states<ul><li><code>genotypeindex</code>: row index starting from 1</li><li><code>genoindex</code>: <em>un-ordered</em> combinations of <code>haploindex</code> in table <code>haplotype</code></li><li><code>genotype</code>: <em>un-ordered</em> combinations of <code>haplotype</code> in table <code>haplotype</code></li></ul></li><li><code>diplotype</code>: ancestral diplotype states<ul><li><code>diplotypeindex</code>: row index starting from 1</li><li><code>diploindex</code>: <em>ordered</em> combinations of <code>haploindex</code> in table <code>haplotype</code></li><li><code>diplotype</code>: <em>ordered</em> combinations of <code>haplotype</code> in table <code>haplotype</code></li></ul></li><li><code>inbredcoef</code>: inbreeding coefficient<ul><li><code>linkagegroup</code>: linkage group ID</li><li><code>offspring</code>: offspring ID</li><li><code>inbredcoef</code>: inbreeding coefficients at all markers in the linkage group for the offspring</li></ul></li><li><code>loglike</code>: log likelihood for each offspring in each chromosome<ul><li><code>linkagegroup</code>: linkage group ID</li><li>1st offpsringID: marginal log likelihood for markers in the chromosome for the offspring    </li><li>...</li><li>last offpsringID</li></ul></li></ul><p>There are 4 more tables: <code>haploprob</code>, <code>genoprob</code>, <code>diploprob</code>, and <code>viterbipath</code>, which will be described in the following.</p><h3 id="ancestryfile-from-forward-backward"><a class="docs-heading-anchor" href="#ancestryfile-from-forward-backward">ancestryfile from forward-backward</a><a id="ancestryfile-from-forward-backward-1"></a><a class="docs-heading-anchor-permalink" href="#ancestryfile-from-forward-backward" title="Permalink"></a></h3><p>In the <code>ancestryfile</code> resulting from the forward-backward HMM algorithm (<code>hmmalg=&quot;forwardbackward&quot;</code>), the main results are saved in tables <code>haploprob</code>, <code>genoprob</code>, and <code>diploprob</code>, and table <code>viterbipath</code> is nothing.</p><ul><li><code>haploprob</code>: haplotype probablity<ul><li><code>linkagegroup</code>: linkage group ID</li><li><code>offspring</code>: offpsring ID</li><li><code>nmarker</code>: number of rows (markers)</li><li><code>nhaplotype</code>: number of columns (haplotypes)</li><li><code>markerindex</code>: row index, index for <code>marker</code> in table <code>foundergeno</code></li><li><code>haplotypeindex</code>: column index, see table <code>haplotype</code>.</li><li><code>haploprob</code>: non-zero haplotype probability, rounded to 4 digits after the decimal place.</li></ul></li><li><code>genoprob</code>: genotype probablity. It is nothing if <code>model = &quot;depmodel&quot;</code><ul><li><code>linkagegroup</code>: linkage group ID</li><li><code>offspring</code>: offpsring ID</li><li><code>nmarker</code>: number of rows (markers)</li><li><code>ngenotype</code>: number of columns (genotypes)</li><li><code>markerindex</code>: row index, index for <code>marker</code> in Table <code>foundergeno</code></li><li><code>genotypeindex</code>: column index, see Table <code>genotype</code>. </li><li><code>genoprob</code>: non-zero genotype probability, rounded to 4 digits after the decimal place.</li></ul></li><li><code>diploprob</code>: diplotype probablity. It is nothing if <code>model = &quot;depmodel&quot;</code><ul><li><code>linkagegroup</code>: linkage group ID</li><li><code>offspring</code>: offpsring ID</li><li><code>nmarker</code>: number of rows (markers)    </li><li><code>ndiplotype</code>: number of columns (diplotypes)</li><li><code>markerindex</code>: row index, index for <code>marker</code> in Table <code>foundergeno</code></li><li><code>diplotypeindex</code>: column index, see Table <code>diplotype</code>. </li><li><code>diploprob</code>: non-zero diplot probability, rounded to 4 digits after the decimal place.</li></ul></li></ul><div class="admonition is-info"><header class="admonition-header">condprob</header><div class="admonition-body"><ul><li>Each row denotes a sparse matrix of posterior probabilities for the offspring in the linkage group. Note that an offspring from a biparental sub-population has at most two non-zero probabilities, despite of many founders for the whole population.</li></ul></div></div><p>For example, consider an offspring produced from 4 inbred founders, the haploprob for 8 markers in chr1 looks like</p><pre><code class="nohighlight hljs">RABBIT, haploprob
linkagegroup, offspring, nmarker, nhaplotype, markerindex, haplotypeindex, haploprob
LG1, offspring1, 8, 4, 3|6|8|1|2|3|4|6|7|1|4|5|6|7|8|2|3|5|7|8, 1|1|1|2|2|2|2|2|2|3|3|3|3|3|3|4|4|4|4|4, 0.01|0.08|0.06|0.85|0.9|0.9|0.82|0.02|0.04|0.15|0.18|0.88|0.9|0.9|0.85|0.1|0.09|0.12|0.06|0.09
...</code></pre><pre><code class="language-julia hljs">using SparseArrays
A = [0.0 0.85 0.15 0.0
     0.0 0.9 0.0 0.1
     0.01 0.9 0.0 0.09
     0.0  0.82 0.18 0.0
     0.0  0.0  0.88 0.12
     0.08  0.02 0.9 0.0
     0.0   0.04 0.9  0.06
     0.06  0.0 0.85 0.09
     ]
B = sparse(A)
Is,Js, Vs = findnz(B)
m, n = size(B)
println(string(&quot;nmarker, nhaplotype = &quot;, (m,n), &quot;\n&quot;,
  &quot;markerindex=&quot;,join(Is,&quot;|&quot;),&quot;\n&quot;,
  &quot;haplotypeindex=&quot;,join(Js,&quot;|&quot;), &quot;\n&quot;,
  &quot;haploprob=&quot;,join(Vs,&quot;|&quot;)))</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">nmarker, nhaplotype = (8, 4)
markerindex=3|6|8|1|2|3|4|6|7|1|4|5|6|7|8|2|3|5|7|8
haplotypeindex=1|1|1|2|2|2|2|2|2|3|3|3|3|3|3|4|4|4|4|4
haploprob=0.01|0.08|0.06|0.85|0.9|0.9|0.82|0.02|0.04|0.15|0.18|0.88|0.9|0.9|0.85|0.1|0.09|0.12|0.06|0.09</code></pre><pre><code class="language-julia hljs">sparse(Is,Js,Vs,m,n)</code></pre><pre class="documenter-example-output"><code class="nohighlight hljs ansi">8×4 SparseArrays.SparseMatrixCSC{Float64, Int64} with 20 stored entries:
  ⋅    0.85  0.15   ⋅ 
  ⋅    0.9    ⋅    0.1
 0.01  0.9    ⋅    0.09
  ⋅    0.82  0.18   ⋅ 
  ⋅     ⋅    0.88  0.12
 0.08  0.02  0.9    ⋅ 
  ⋅    0.04  0.9   0.06
 0.06   ⋅    0.85  0.09</code></pre><h3 id="ancestryfile-from-Viterbi"><a class="docs-heading-anchor" href="#ancestryfile-from-Viterbi">ancestryfile from Viterbi</a><a id="ancestryfile-from-Viterbi-1"></a><a class="docs-heading-anchor-permalink" href="#ancestryfile-from-Viterbi" title="Permalink"></a></h3><p>In the <code>ancestryfile</code> resulting from the Viterbi HMM algorithm (hmmalg=viterbi), the main results are saved in table <code>viterbipath</code>, and tables <code>haploprob</code>, <code>genoprob</code>, and <code>diploprob</code> are nothing.</p><div class="admonition is-info"><header class="admonition-header">Table viterbipath</header><div class="admonition-body"><ul><li>Table viterbipath has two columns: <code>individual</code> and <code>viterbipath</code>.</li><li>Each row denotes the viterbi paths for all linkage groups in an offspring; paths between linkage groups are joined by &quot;|&quot;.</li><li>The viterbi path for a linkage group in an offspring is in the form of m(1)-s(1)-m(2)-s(2)...m(k)-s(k), where m(1), m(2),..., m(k) denote the marker indices, and s(1), s(2), ..., s(k) denote the indices of HMM hidden states. Note that it always holds that m(1)=1 and m(k) = 1 + number of markers in the linkage group. The i-th segment has state s(i) that starts at m(i) and ends at m(i+1) -1.  </li><li>If model = depmodel, s(i) refers to haplotypeindex, and it refers to diplotypeindex for model = jointmodel or indepmodel.</li></ul></div></div><p>For example, consider an offspring produced from 4 inbred parents that has two linkage groups.</p><pre><code class="nohighlight hljs">RABBIT, viterbipath
individual, viterbipath
offspring1, 1-12-96-11-101-16-122|1-16-143
...</code></pre><p>The viterbi path for the 1st linkage group is 1-12-96-11-101-16-122, meaning that segment 1 has state 12 from marker indices 1 to 95, segment 2 has state 11 from marker indices 96 to 100, and segment 3 has state 16 from marker indices 101 to 121.</p><p>The viterbi path for the 2nd linkage group is 1-16-143, a single segment has state 16 for all the 142 markers.</p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../prepare_input/prepare_convert/">« Prepare by converting </a><a class="docs-footer-nextpage" href="../key_options/">Key options »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Tuesday 7 January 2025 16:05">Tuesday 7 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
