<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Statistical framework · RABBIT</title><meta name="title" content="Statistical framework · RABBIT"/><meta property="og:title" content="Statistical framework · RABBIT"/><meta property="twitter:title" content="Statistical framework · RABBIT"/><meta name="description" content="Documentation for RABBIT."/><meta property="og:description" content="Documentation for RABBIT."/><meta property="twitter:description" content="Documentation for RABBIT."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">RABBIT</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2"><span class="docs-label">Guide</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../run_rabbit/">Run RABBIT</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Prepare Inputs</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../prepare_input/prepare_genofile/">Prepare genofile</a></li><li><a class="tocitem" href="../prepare_input/prepare_pedinfo_designcode/">Prepare pedinfo: designcode</a></li><li><a class="tocitem" href="../prepare_input/prepare_pedinfo_pedfile/">Prepare pedinfo: pedfile</a></li><li><a class="tocitem" href="../prepare_input/prepare_phenofile/">Prepare phenofile</a></li><li><a class="tocitem" href="../prepare_input/prepare_convert/">Prepare by converting </a></li></ul></li><li><a class="tocitem" href="../understand_output/">Understand outputs</a></li><li><a class="tocitem" href="../key_options/">Key options</a></li><li class="is-active"><a class="tocitem" href>Statistical framework</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox"/><label class="tocitem" for="menuitem-3"><span class="docs-label">Pipeline</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../pipeline/s0_pipeline_overview/">Overview</a></li><li><a class="tocitem" href="../pipeline/s1_magicsimulate/">S1 MagicSimulate</a></li><li><a class="tocitem" href="../pipeline/s2_magicfilter/">S2 MagicFilter</a></li><li><a class="tocitem" href="../pipeline/s3_magiccall/">S3 MagicCall</a></li><li><a class="tocitem" href="../pipeline/s4_magicmap/">S4 MagicMap</a></li><li><a class="tocitem" href="../pipeline/s5_magicimpute/">S5 MagicImpute</a></li><li><a class="tocitem" href="../pipeline/s6_magicreconstruct/">S6 MagicReconstruct</a></li><li><a class="tocitem" href="../pipeline/s7_magicscan/">S7 MagicScan</a></li></ul></li><li><a class="tocitem" href="../rabbit_cli/">CLI</a></li><li><a class="tocitem" href="../rabbit_api/">API</a></li><li><a class="tocitem" href="../ref/">Citations</a></li></ul><div class="docs-version-selector field has-addons visible"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"><option value="#" selected="selected">v1.0.0</option></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Guide</a></li><li class="is-active"><a href>Statistical framework</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Statistical framework</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Statistical-framework"><a class="docs-heading-anchor" href="#Statistical-framework">Statistical framework</a><a id="Statistical-framework-1"></a><a class="docs-heading-anchor-permalink" href="#Statistical-framework" title="Permalink"></a></h1><p>See <a href="https://doi.org/10.1534/genetics.114.163006"><code>Zheng et al. 2014</code></a>, <a href="https://doi.org/10.1534/g3.114.016154"><code>Zheng 2015</code></a> , <a href="https://doi.org/10.1534/genetics.115.177873"><code>Zheng et al. 2015</code></a>, and <a href="https://doi.org/10.1534/G3.118.200340"><code>Zheng et al. 2018</code></a> for the detailed description on the Hidden Markov Model (HMM) framework of RABBIT.  See <a href="https://doi.org/10.1534/genetics.118.300885"><code>Zheng et al. 2018(2)</code></a>  and <code>Zheng et al. 2024</code> for the description on the algorithm of genotype imputation.  See <a href="https://doi.org/10.1534/genetics.119.302229"><code>Zheng et al. 2019</code></a> and <code>Zheng et al. 2024(2)</code> for the description on the algorithm of genetic map construction. </p><p>The RABBIT HMM framework consists of two basic components: hidden Markov process and genotype data model. </p><h3 id="Ancestral-origin-process"><a class="docs-heading-anchor" href="#Ancestral-origin-process">Ancestral origin process</a><a id="Ancestral-origin-process-1"></a><a class="docs-heading-anchor-permalink" href="#Ancestral-origin-process" title="Permalink"></a></h3><p>The hidden Markov process refers to the prior ancestral origin process, describing how ancestral origins change along two homologous chromosomes in a diploid offspring.</p><p>RABBIT has a keyarg <code>model</code> for specifying the dependence of the prior ancestral origin processes between two homologous chromosomes. It must be &quot;depmodel&quot;, &quot;indepmodel&quot;, or &quot;jointmodel&quot;, denoting complete dependence, complete independence, or intermediate dependence, respectively. RABBIT uses the general &quot;jointmodel&quot; by default. It is recommended for <code>magicimpute</code> to use &quot;depmodel&quot; for almost homozygous populations, which would be much faster than the default &quot;jointmodel&quot;.</p><h3 id="Genotype-data-model"><a class="docs-heading-anchor" href="#Genotype-data-model">Genotype data model</a><a id="Genotype-data-model-1"></a><a class="docs-heading-anchor-permalink" href="#Genotype-data-model" title="Permalink"></a></h3><p>The Genotype data model describes the (emission) probability of observed genotypic data given hidden ancestral origin state, and it varies with genotype format. See [<code>Zheng et al. 2024</code>] for the detailed description on the data model. </p><h4 id="Discrete-genotype-(GT)"><a class="docs-heading-anchor" href="#Discrete-genotype-(GT)">Discrete genotype (GT)</a><a id="Discrete-genotype-(GT)-1"></a><a class="docs-heading-anchor-permalink" href="#Discrete-genotype-(GT)" title="Permalink"></a></h4><p>The data model for &quot;GT&quot; has a parameter describing the allelic error rate, that is, the probability of an error occurring on one allele. If an error occurs on an allele, it will result in the other allele.</p><p>RABBIT introduces two likelihood parameters for &quot;GT&quot;: <code>foundererror</code> and <code>offspringerror</code>, denoting the allelic error rates for founders and offspring, respectively.</p><h4 id="Allelic-depeth-(AD)"><a class="docs-heading-anchor" href="#Allelic-depeth-(AD)">Allelic depeth (AD)</a><a id="Allelic-depeth-(AD)-1"></a><a class="docs-heading-anchor-permalink" href="#Allelic-depeth-(AD)" title="Permalink"></a></h4><p>Sequence reads are assumed to be generated by two steps: (1) true genotypes are mis-aligned using the random allele model with the allelic error rates <code>foundererror</code> and <code>offspringerror</code>, and (2) conditional on mis-aligned genotypes, sequence reads are sampled with parameters <code>seqerror</code>, <code>allelebalancemean</code>, <code>allelebalancedisperse</code>, and <code>alleledropout</code>.</p><p>We extend the binomial model such that a heterozygous genotype is modeled by a 0- and 1- inflated beta-binomial distribution. Conditional on mis-aligned genotypes, we model sequence reads as follows</p><p class="math-container">\[\begin{aligned}
P(r0,r1|0/0) &amp;= B(r0|n, 1-seqerror) \\
P(r0,r1|0/1) &amp;= \text{Inf-BetaBin}(r0|n, seqerror, allelebalancemean, allelebalancedisperse, alleledropout) \\
P(r0,r1|1/1) &amp;= B(r0|n, seqerror)
\end{aligned}\]</p><p>where r0 is the read count for allele 0, r1 is the read count for allele 1, <span>$n=r0+r1$</span>, and <span>$B(r0|n,p)$</span> is the binomial distribution. Conditional on heterozygous genotype, the allele 0 is dropped out and <code>P(r0,r1|0/1)</code> is given by <code>P(r0,r1|1/1)</code> with  probability <code>alleledropout/2</code>, the allele 1 is dropped out and <code>P(r0,r1|0/1)</code> is given by <code>P(r0,r1|0/0)</code> with probability <code>alleledropout/2</code>, and otherwise <code>P(r0,r1|0/1)</code> follows a beta-binomial distribution <span>$BetaBin(r0|n, \alpha, \beta)$</span>. The beta-binomial distribution can be regarded as the compound distribution that results from  <span>$r0 \sim B(r0|n,p)$</span> with <span>$p \sim Beta(\alpha, \beta)$</span>, where <code>p</code> is the allelic balance for an offspring.  We reparameterize <span>$\alpha$</span> and <span>$\beta$</span>  as follows</p><p class="math-container">\[\begin{aligned}
\alpha &amp;= allelebalancemean/allelebalancedisperse \\
\beta &amp;= (1-allelebalancemean)/allelebalancedisperse
\end{aligned}\]</p><div class="admonition is-info"><header class="admonition-header">Sequence read model</header><div class="admonition-body"><ul><li><span>$BetaBin(n, \alpha, \beta) \rightarrow B(n, allelebalancemean)$</span>, as <span>$allelebalancedisperse \rightarrow 0$</span>.</li><li>The sequence read model reduces to the binomial model as  as <span>$allelebalancedisperse \rightarrow 0$</span> and <span>$allelebalancemean \rightarrow 0.5$</span></li><li><span>$allelebalancemean - 0.5$</span> measures allelic balance bias, and <span>$allelebalancedisperse$</span> measures allelic balance overdispersion.</li></ul></div></div><h3 id="Referneces"><a class="docs-heading-anchor" href="#Referneces">Referneces</a><a id="Referneces-1"></a><a class="docs-heading-anchor-permalink" href="#Referneces" title="Permalink"></a></h3><p>Zheng, Chaozhi, Martin P Boer, and Fred A Van Eeuwijk. 2014. “A General Modeling Framework for Genome Ancestral Origins in Multiparental Populations.” Genetics 198 (1): 87–101. https://doi.org/10.1534/genetics.114.163006. </p><p>———. 2015. “Reconstruction of Genome Ancestry Blocks in Multiparental Populations.” Genetics 200 (4): 1073–87. https://doi.org/10.1534/genetics.115.177873.</p><p>———. 2018. “Recursive Algorithms for Modeling Genomic Ancestral Origins in a Fixed Pedigree.” G3 Genes|Genomes|Genetics 8 (10): 3231–45. https://doi.org/10.1534/G3.118.200340.</p><p>———. 2018(2). &quot;Accurate Genotype Imputation in Multiparental Populations from Low-Coverage Sequence&quot;. Genetics 210 (1): 71-82. https://doi.org/10.1534/genetics.118.300885</p><p>———. 2019. &quot;Construction of Genetic Linkage Maps in Multiparental Populations&quot;. Genetics 212 (4): 1031-1044. https://doi.org/10.1534/genetics.119.302229</p><p>Zheng et al. 2024. “Genotype imputation in connected multiparental populations.”  In preparation.</p><p>Zheng et al. 2024(2). &quot;Efficient consensus map construction in connected multiparental populations.”  In preparation. </p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../key_options/">« Key options</a><a class="docs-footer-nextpage" href="../pipeline/s0_pipeline_overview/">Overview »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Tuesday 7 January 2025 16:05">Tuesday 7 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
