<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>S2 MagicFilter · RABBIT</title><meta name="title" content="S2 MagicFilter · RABBIT"/><meta property="og:title" content="S2 MagicFilter · RABBIT"/><meta property="twitter:title" content="S2 MagicFilter · RABBIT"/><meta name="description" content="Documentation for RABBIT."/><meta property="og:description" content="Documentation for RABBIT."/><meta property="twitter:description" content="Documentation for RABBIT."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">RABBIT</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox"/><label class="tocitem" for="menuitem-2"><span class="docs-label">Guide</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../run_rabbit/">Run RABBIT</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Prepare Inputs</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../prepare_input/prepare_genofile/">Prepare genofile</a></li><li><a class="tocitem" href="../../prepare_input/prepare_pedinfo_designcode/">Prepare pedinfo: designcode</a></li><li><a class="tocitem" href="../../prepare_input/prepare_pedinfo_pedfile/">Prepare pedinfo: pedfile</a></li><li><a class="tocitem" href="../../prepare_input/prepare_phenofile/">Prepare phenofile</a></li><li><a class="tocitem" href="../../prepare_input/prepare_convert/">Prepare by converting </a></li></ul></li><li><a class="tocitem" href="../../understand_output/">Understand outputs</a></li><li><a class="tocitem" href="../../key_options/">Key options</a></li><li><a class="tocitem" href="../../statistical_framework/">Statistical framework</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox" checked/><label class="tocitem" for="menuitem-3"><span class="docs-label">Pipeline</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../s0_pipeline_overview/">Overview</a></li><li><a class="tocitem" href="../s1_magicsimulate/">S1 MagicSimulate</a></li><li class="is-active"><a class="tocitem" href>S2 MagicFilter</a></li><li><a class="tocitem" href="../s3_magiccall/">S3 MagicCall</a></li><li><a class="tocitem" href="../s4_magicmap/">S4 MagicMap</a></li><li><a class="tocitem" href="../s5_magicimpute/">S5 MagicImpute</a></li><li><a class="tocitem" href="../s6_magicreconstruct/">S6 MagicReconstruct</a></li><li><a class="tocitem" href="../s7_magicscan/">S7 MagicScan</a></li></ul></li><li><a class="tocitem" href="../../rabbit_cli/">CLI</a></li><li><a class="tocitem" href="../../rabbit_api/">API</a></li><li><a class="tocitem" href="../../ref/">Citations</a></li></ul><div class="docs-version-selector field has-addons visible"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"><option value="#" selected="selected">v1.0.0</option></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Pipeline</a></li><li class="is-active"><a href>S2 MagicFilter</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>S2 MagicFilter</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="S2-MagicFilter"><a class="docs-heading-anchor" href="#S2-MagicFilter">S2 MagicFilter</a><a id="S2-MagicFilter-1"></a><a class="docs-heading-anchor-permalink" href="#S2-MagicFilter" title="Permalink"></a></h1><h3 id="Data-filtering"><a class="docs-heading-anchor" href="#Data-filtering">Data filtering</a><a id="Data-filtering-1"></a><a class="docs-heading-anchor-permalink" href="#Data-filtering" title="Permalink"></a></h3><div class="admonition is-success"><header class="admonition-header">filter markers</header><div class="admonition-body"><p>It is recommended to first run <a href="../../rabbit_api/#MagicBase.vcffilter"><code>MagicBase.vcffilter</code></a> for fast online marker filtering. </p></div></div><p>The <code>magicfilter</code> sequentially performs the following:</p><ol><li>Filter for subpopulations with size &gt;= <code>min_subpop</code>. By default, <code>min_subpop=1</code> (i.e. not performed).</li><li>Remove founders (and their offspring) if their numbers of progeny &lt; <code>min_nprogeny</code>. By default, <code>min_nprogeny=1</code> (i.e., not performed).</li><li>Test monomorphic for each subpopulation at each marker.</li><li>Filter for markers with <code>snp_missfilter(f,o)</code> for founder missing fraction <code>f</code> and offspring missing fraction <code>o</code>.</li><li>Filter for markers with minor allele frequency &lt;= <code>snp_minmaf</code>.  </li><li>Filter for offspring with missing fraction &lt;= <code>offspring_maxmiss</code>.</li><li>Remove duplicated offspring (<code>isfilterdupe = true</code>) by keeping one of them with the least missing values; a pair of offspring are duplicated if their allelic correlation &gt;= <code>offspring_maxcorr</code>.</li></ol><pre><code class="language-julia hljs"># code for Julia
using MagicFilter  
cd(@__DIR__)       
genofile = outstem*&quot;_magicsimulate_geno.vcf.gz&quot;
pedfile = outstem*&quot;_magicsimulate_ped.csv&quot;
magicfilter(genofile,pedfile;    
    snp_minmaf = 0.05, 
    snp_missfilter = (f,o) -&gt; o &lt;= 0.9 || f &lt;= 0.1, 
    offspring_maxmiss = 0.95,  
    isfilterdupe = true,  # default false
    outstem
)</code></pre><pre><code class="language-bash hljs"># code for Linux shell. 
# For Window CMD, replace multiline key \ by  ^, and replace comment-key # by ::
# For CLI, snp_missfilter = (f,o) -&gt; o &lt;= snp_maxomiss || f &lt;= or_snp_maxfmiss
julia rabbit_magicfilter.jl -g example_magicsimulate_geno.vcf.gz \
    -p example_magicsimulate_ped.csv \
    --snp_minmaf 0.05 --snp_mono2miss true \
    --snp_maxomiss 0.9 \
    --or_snp_maxfmiss 0.1 \
    --offspring_maxmiss 0.95 \
    --isfilterdupe true \
    -o example</code></pre><h4 id="Output-files"><a class="docs-heading-anchor" href="#Output-files">Output files</a><a id="Output-files-1"></a><a class="docs-heading-anchor-permalink" href="#Output-files" title="Permalink"></a></h4><table><tr><th style="text-align: left">outfile</th><th style="text-align: left">Description</th></tr><tr><td style="text-align: left">outstem*&quot;_magicfilter.log&quot;</td><td style="text-align: left">log file</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfilter_geno.vcf.gz&quot;</td><td style="text-align: left">filtered genofile for downstream analysis</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfilter_ped.csv&quot;</td><td style="text-align: left">filtered pedfile for downstream analysis</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfilter_ped.png&quot;</td><td style="text-align: left">plot for filtered pedfile</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_subpop_size.csv&quot;</td><td style="text-align: left">size of each subpopulation</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_subpop_size.png&quot;</td><td style="text-align: left">plot distribution of subpopulation size</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_founder_nprogeny.csv&quot;</td><td style="text-align: left">number of progeny for each founder</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_founder_change.csv&quot;</td><td style="text-align: left">changes in founder genotypes</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_ind_missing.csv&quot;</td><td style="text-align: left">missing fraction for founders and offspring</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_ind_missing.png&quot;</td><td style="text-align: left">plot distribution of missing fraction among individuals</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_snp_sum.csv&quot;</td><td style="text-align: left">summary of monomorphic test</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_snp_MAF.png&quot;</td><td style="text-align: left">plot distribution of MAF</td></tr><tr><td style="text-align: left">outstem*&quot;_magicfiler_snp_missing.png&quot;</td><td style="text-align: left">plot distribution of missing fraction among markers</td></tr></table><h3 id="Output:-empirical-CDF-of-MAF"><a class="docs-heading-anchor" href="#Output:-empirical-CDF-of-MAF">Output: empirical CDF of MAF</a><a id="Output:-empirical-CDF-of-MAF-1"></a><a class="docs-heading-anchor-permalink" href="#Output:-empirical-CDF-of-MAF" title="Permalink"></a></h3><p><code>outstem*&quot;_magicfilter_snp_MAF.png&quot;</code> plots the empirical cumulative distribution function (CDF) of MAF per marker. See <code>outstem*&quot;_magicfilter_snp_sum.csv&quot;</code> for the exact results. The MAF is calculated among offspring for each marker. The red dot indicates the fraction of removed markers for the given threshold <code>snp_minmaf = 0.05</code>.  </p><p><img src="../example_magicfilter_snp_MAF.png" alt/></p><h3 id="Output:-empirical-CDF-of-missing-fraction-per-marker"><a class="docs-heading-anchor" href="#Output:-empirical-CDF-of-missing-fraction-per-marker">Output: empirical CDF of missing fraction per marker</a><a id="Output:-empirical-CDF-of-missing-fraction-per-marker-1"></a><a class="docs-heading-anchor-permalink" href="#Output:-empirical-CDF-of-missing-fraction-per-marker" title="Permalink"></a></h3><p><code>outstem*&quot;_magicfilter_snp_missing.png&quot;</code> plots the empirical CDF of missing fraction per marker for founders and offspring, respectively. See  <code>outstem*&quot;_magicfilter_snp_sum.csv&quot;</code> for the exact results. </p><p><img src="../example_magicfilter_snp_missing.png" alt/></p><h3 id="Output:-empirical-CDF-of-missing-fraction-per-individual"><a class="docs-heading-anchor" href="#Output:-empirical-CDF-of-missing-fraction-per-individual">Output: empirical CDF of missing fraction per individual</a><a id="Output:-empirical-CDF-of-missing-fraction-per-individual-1"></a><a class="docs-heading-anchor-permalink" href="#Output:-empirical-CDF-of-missing-fraction-per-individual" title="Permalink"></a></h3><p><code>outstem*&quot;_magicfilter_ind_missing.png&quot;</code> plots the empirical CDF of missing fraction per individual. See  <code>outstem*&quot;_magicfilter_ind_missing.csv&quot;</code> for the exact results.  The red dot indicates 1 - fraction of removed markers for the given threshold <code>offspring_maxmiss</code>. </p><p><img src="../example_magicfilter_ind_missing.png" alt/></p><h3 id="Output:-offspring-duplicate"><a class="docs-heading-anchor" href="#Output:-offspring-duplicate">Output: offspring duplicate</a><a id="Output:-offspring-duplicate-1"></a><a class="docs-heading-anchor-permalink" href="#Output:-offspring-duplicate" title="Permalink"></a></h3><p><code>outstem*&quot;_magicfilter_ind_correlation.png&quot;</code> gives the heatmap of allelic correlation for each pair of offspring. A pair of offspring is assumed to be duplicated if their correlation &gt; <code>offspring_maxmiss</code>. By default, <code>offspring_maxmiss = 0.99</code>.  These output files are produced only if <code>isfilterdupe = true</code>. </p><p><img src="../example_magicfilter_ind_correlation.png" alt/></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../s1_magicsimulate/">« S1 MagicSimulate</a><a class="docs-footer-nextpage" href="../s3_magiccall/">S3 MagicCall »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Tuesday 7 January 2025 16:05">Tuesday 7 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
