<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>API · RABBIT</title><meta name="title" content="API · RABBIT"/><meta property="og:title" content="API · RABBIT"/><meta property="twitter:title" content="API · RABBIT"/><meta name="description" content="Documentation for RABBIT."/><meta property="og:description" content="Documentation for RABBIT."/><meta property="twitter:description" content="Documentation for RABBIT."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">RABBIT</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox"/><label class="tocitem" for="menuitem-2"><span class="docs-label">Guide</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../run_rabbit/">Run RABBIT</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Prepare Inputs</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../prepare_input/prepare_genofile/">Prepare genofile</a></li><li><a class="tocitem" href="../prepare_input/prepare_pedinfo_designcode/">Prepare pedinfo: designcode</a></li><li><a class="tocitem" href="../prepare_input/prepare_pedinfo_pedfile/">Prepare pedinfo: pedfile</a></li><li><a class="tocitem" href="../prepare_input/prepare_phenofile/">Prepare phenofile</a></li><li><a class="tocitem" href="../prepare_input/prepare_convert/">Prepare by converting </a></li></ul></li><li><a class="tocitem" href="../understand_output/">Understand outputs</a></li><li><a class="tocitem" href="../key_options/">Key options</a></li><li><a class="tocitem" href="../statistical_framework/">Statistical framework</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox"/><label class="tocitem" for="menuitem-3"><span class="docs-label">Pipeline</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../pipeline/s0_pipeline_overview/">Overview</a></li><li><a class="tocitem" href="../pipeline/s1_magicsimulate/">S1 MagicSimulate</a></li><li><a class="tocitem" href="../pipeline/s2_magicfilter/">S2 MagicFilter</a></li><li><a class="tocitem" href="../pipeline/s3_magiccall/">S3 MagicCall</a></li><li><a class="tocitem" href="../pipeline/s4_magicmap/">S4 MagicMap</a></li><li><a class="tocitem" href="../pipeline/s5_magicimpute/">S5 MagicImpute</a></li><li><a class="tocitem" href="../pipeline/s6_magicreconstruct/">S6 MagicReconstruct</a></li><li><a class="tocitem" href="../pipeline/s7_magicscan/">S7 MagicScan</a></li></ul></li><li><a class="tocitem" href="../rabbit_cli/">CLI</a></li><li class="is-active"><a class="tocitem" href>API</a><ul class="internal"><li><a class="tocitem" href="#Contents"><span>Contents</span></a></li><li><a class="tocitem" href="#Index"><span>Index</span></a></li><li><a class="tocitem" href="#Public-Interface"><span>Public Interface</span></a></li></ul></li><li><a class="tocitem" href="../ref/">Citations</a></li></ul><div class="docs-version-selector field has-addons visible"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"><option value="#" selected="selected">v1.0.0</option></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>API</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>API</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="RABBIT-interface-for-julia-REPL"><a class="docs-heading-anchor" href="#RABBIT-interface-for-julia-REPL">RABBIT interface for julia REPL</a><a id="RABBIT-interface-for-julia-REPL-1"></a><a class="docs-heading-anchor-permalink" href="#RABBIT-interface-for-julia-REPL" title="Permalink"></a></h1><p>Documentation for  RABBIT&#39;s public interface.</p><h2 id="Contents"><a class="docs-heading-anchor" href="#Contents">Contents</a><a id="Contents-1"></a><a class="docs-heading-anchor-permalink" href="#Contents" title="Permalink"></a></h2><ul><li><a href="#RABBIT-interface-for-julia-REPL">RABBIT interface for julia REPL</a></li><li class="no-marker"><ul><li><a href="#Contents">Contents</a></li><li><a href="#Index">Index</a></li><li><a href="#Public-Interface">Public Interface</a></li></ul></li></ul><h2 id="Index"><a class="docs-heading-anchor" href="#Index">Index</a><a id="Index-1"></a><a class="docs-heading-anchor-permalink" href="#Index" title="Permalink"></a></h2><ul><li><a href="#MagicBase"><code>MagicBase</code></a></li><li><a href="#MagicCall"><code>MagicCall</code></a></li><li><a href="#MagicFilter"><code>MagicFilter</code></a></li><li><a href="#MagicImpute"><code>MagicImpute</code></a></li><li><a href="#MagicMap"><code>MagicMap</code></a></li><li><a href="#MagicReconstruct"><code>MagicReconstruct</code></a></li><li><a href="#MagicScan"><code>MagicScan</code></a></li><li><a href="#MagicSimulate"><code>MagicSimulate</code></a></li><li><a href="#MagicBase.DesignInfo"><code>MagicBase.DesignInfo</code></a></li><li><a href="#MagicBase.JuncDist"><code>MagicBase.JuncDist</code></a></li><li><a href="#MagicBase.MagicAncestry"><code>MagicBase.MagicAncestry</code></a></li><li><a href="#MagicBase.MagicGeno"><code>MagicBase.MagicGeno</code></a></li><li><a href="#MagicBase.MagicPed"><code>MagicBase.MagicPed</code></a></li><li><a href="#Pedigrees.Pedigree"><code>Pedigrees.Pedigree</code></a></li><li><a href="#MagicBase.animcondprob"><code>MagicBase.animcondprob</code></a></li><li><a href="#MagicBase.array_extract_pedfile"><code>MagicBase.array_extract_pedfile</code></a></li><li><a href="#MagicBase.arrayfile2vcf"><code>MagicBase.arrayfile2vcf</code></a></li><li><a href="#MagicBase.formmagicgeno"><code>MagicBase.formmagicgeno</code></a></li><li><a href="#MagicBase.formmagicped"><code>MagicBase.formmagicped</code></a></li><li><a href="#MagicBase.hapmap2vcf"><code>MagicBase.hapmap2vcf</code></a></li><li><a href="#MagicBase.merge_arrayfiles"><code>MagicBase.merge_arrayfiles</code></a></li><li><a href="#MagicBase.merge_pedfiles"><code>MagicBase.merge_pedfiles</code></a></li><li><a href="#MagicBase.merge_vcffiles"><code>MagicBase.merge_vcffiles</code></a></li><li><a href="#MagicBase.parsebreedped"><code>MagicBase.parsebreedped</code></a></li><li><a href="#MagicBase.parsedesign"><code>MagicBase.parsedesign</code></a></li><li><a href="#MagicBase.pedfile_designcode2ped"><code>MagicBase.pedfile_designcode2ped</code></a></li><li><a href="#MagicBase.plotcondprob"><code>MagicBase.plotcondprob</code></a></li><li><a href="#MagicBase.plotmagicped"><code>MagicBase.plotmagicped</code></a></li><li><a href="#MagicBase.plotmarkermap"><code>MagicBase.plotmarkermap</code></a></li><li><a href="#MagicBase.plotrecombreak"><code>MagicBase.plotrecombreak</code></a></li><li><a href="#MagicBase.rabbitgeno_mma2jl"><code>MagicBase.rabbitgeno_mma2jl</code></a></li><li><a href="#MagicBase.rabbitped_mma2jl"><code>MagicBase.rabbitped_mma2jl</code></a></li><li><a href="#MagicBase.readmagicancestry"><code>MagicBase.readmagicancestry</code></a></li><li><a href="#MagicBase.readmagicped"><code>MagicBase.readmagicped</code></a></li><li><a href="#MagicBase.resetmap"><code>MagicBase.resetmap</code></a></li><li><a href="#MagicBase.savegenodata"><code>MagicBase.savegenodata</code></a></li><li><a href="#MagicBase.savemagicancestry"><code>MagicBase.savemagicancestry</code></a></li><li><a href="#MagicBase.savemagicgeno"><code>MagicBase.savemagicgeno</code></a></li><li><a href="#MagicBase.savemagicped"><code>MagicBase.savemagicped</code></a></li><li><a href="#MagicBase.vcf_extract_pedfile"><code>MagicBase.vcf_extract_pedfile</code></a></li><li><a href="#MagicBase.vcf_pad_samples"><code>MagicBase.vcf_pad_samples</code></a></li><li><a href="#MagicBase.vcffilter"><code>MagicBase.vcffilter</code></a></li><li><a href="#MagicCall.magiccall"><code>MagicCall.magiccall</code></a></li><li><a href="#MagicFilter.magicfilter"><code>MagicFilter.magicfilter</code></a></li><li><a href="#MagicFilter.magicfilter!"><code>MagicFilter.magicfilter!</code></a></li><li><a href="#MagicImpute.magicimpute"><code>MagicImpute.magicimpute</code></a></li><li><a href="#MagicImpute.magicimpute!"><code>MagicImpute.magicimpute!</code></a></li><li><a href="#MagicMap.magicmap"><code>MagicMap.magicmap</code></a></li><li><a href="#MagicPrior.magicorigin"><code>MagicPrior.magicorigin</code></a></li><li><a href="#MagicPrior.magicprior"><code>MagicPrior.magicprior</code></a></li><li><a href="#MagicReconstruct.magicreconstruct"><code>MagicReconstruct.magicreconstruct</code></a></li><li><a href="#MagicReconstruct.magicreconstruct!"><code>MagicReconstruct.magicreconstruct!</code></a></li><li><a href="#MagicScan.magicscan"><code>MagicScan.magicscan</code></a></li><li><a href="#MagicSimulate.magicsimulate"><code>MagicSimulate.magicsimulate</code></a></li><li><a href="#MagicSimulate.simfhaplo"><code>MagicSimulate.simfhaplo</code></a></li><li><a href="#Pedigrees.orderped"><code>Pedigrees.orderped</code></a></li><li><a href="#Pedigrees.plotped"><code>Pedigrees.plotped</code></a></li><li><a href="#Pedigrees.readped"><code>Pedigrees.readped</code></a></li><li><a href="#Pedigrees.saveped"><code>Pedigrees.saveped</code></a></li></ul><h2 id="Public-Interface"><a class="docs-heading-anchor" href="#Public-Interface">Public Interface</a><a id="Public-Interface-1"></a><a class="docs-heading-anchor-permalink" href="#Public-Interface" title="Permalink"></a></h2><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Pedigrees.Pedigree" href="#Pedigrees.Pedigree"><code>Pedigrees.Pedigree</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">Pedigree{T}</code></pre><p>immutable struct that stores information of a pedigree. See also <a href="#Pedigrees.readped"><code>readped</code></a>.</p><p>The Pedigree fields include <code>nfounder</code>, <code>generation</code>, <code>member</code>, <code>gender</code>, <code>mother</code>, and <code>father</code>. The mothers and fathers of founders must be denoted by 0.</p><pre><code class="nohighlight hljs">Pedigree(nfounder,member,mother,father,gender,generation)</code></pre><p>inner constructor. The gender of each member must be &quot;notapplicable&quot;, &quot;female&quot;, or &quot;male&quot;.</p><pre><code class="nohighlight hljs">Pedigree(df::DataFrame)</code></pre><p>extenral constructor. The dataframe must have columns <code>member</code>, <code>mother</code>, and <code>father</code>. A gender column is optional. The constructor calculates the generation for each member, which is the max path length from founders to the member. The constructor orders pedigree by calling <a href="#Pedigrees.orderped"><code>orderped</code></a>.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Pedigrees.orderped" href="#Pedigrees.orderped"><code>Pedigrees.orderped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">orderped(ped::Pedigree)</code></pre><p>order pedigree members so that parents always come first.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Pedigrees.readped" href="#Pedigrees.readped"><code>Pedigrees.readped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readped(pedfile::AbstractString, delim=&#39;,&#39;, commentstring=&quot;##&quot;)</code></pre><p>read a pedigree from pedfile, ignoring the lines beginning with commentstring.</p><p>The pedigree file  must contain columns: member, mother, and father. The mothers and fathers of founders must be denoted by 0. The value in the optional gender column must  be &quot;notapplicable&quot;, &quot;female&quot; or &quot;male&quot;.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Pedigrees.saveped" href="#Pedigrees.saveped"><code>Pedigrees.saveped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">saveped(filename::AbstractString,ped::Pedigree)</code></pre><p>save pedigree into a csv file. The file contains columns: member, mother, father, gender, generation.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="Pedigrees.plotped" href="#Pedigrees.plotped"><code>Pedigrees.plotped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plot(ped::Pedigree)</code></pre><p>plot a pedigree. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicPrior.magicorigin" href="#MagicPrior.magicorigin"><code>MagicPrior.magicorigin</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicorigin(pedigree::Pedigree; kwargs...)</code></pre><p>calculate expected ancestral probabilities and junction densities  in pedigree.</p><p><strong>Keyword Argument</strong></p><p><code>memberlist::Union{Nothing,AbstractVector}=nothing</code>: a list of pedigree members. By default, memberlist contains the last pedigree member.</p><p><code>isautosome::Bool=true</code>: autosome rather than sex chromosome,</p><p><code>isconcise::Bool=false</code>: if false, calculate quantities: phi12, R(a)(or R^m), R(b)(or R^p), ρ, J1112, J1121, J1122, J1211, J1213, J1222, J1232. If true, calculate only the first 4 quantities.</p><p><code>isfglexch::Bool=false</code>: if true, assume that  founder genotype lables (FGLs) are exchangeable.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicPrior.magicprior" href="#MagicPrior.magicprior"><code>MagicPrior.magicprior</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicprior(pedigree::Pedigree,founderfgl::AbstractMatrix; kwargs...)</code></pre><p>calculate prior distribution of recombination breakpoints in pedigree. </p><p>For each pedigree member in <code>memberlist</code>, return continutuous Markov process parameter values: initial probability vector and rate matrix under three models &quot;depmodel&quot;, &quot;indepmodel&quot;, and &quot;jointmodel&quot;, a relationship between the two ancestral processes along each of two homologous chromosomes.</p><p><strong>Position arguments</strong></p><p><code>pedigree::Pedigree</code>: pedigree struct. </p><p><code>isfounderinbred::Bool</code>: if true, founders are inbred, and otherwise outbred.</p><p><strong>Keyword Arguments</strong></p><p><code>memberlist::Union{Nothing,AbstractVector}=nothing</code>: a list of pedigree members.</p><p><code>isautosome::Bool=true</code>: autosome rather than sex chromosome,</p><p><code>isfglexch::Bool=false</code>: not assume that  founder genotype lables (FGLs) are exchangeable.</p><p><code>isconcise::Bool=false</code>: if isconcise is true, the parameter values for &quot;jointmodel&quot; are not calculated.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase" href="#MagicBase"><code>MagicBase</code></a> — <span class="docstring-category">Module</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicBase</code></pre><p>a package for basic data structures and functions for genetic analysis in connected multiparental populations. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.JuncDist" href="#MagicBase.JuncDist"><code>MagicBase.JuncDist</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">JuncDist</code></pre><p>struct that stores junction information, that is, the prior distribution of recombination breakpoints. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.DesignInfo" href="#MagicBase.DesignInfo"><code>MagicBase.DesignInfo</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">DesignInfo</code></pre><p>mutable struct that stores design information for a subpopulation. See also <a href="#MagicBase.parsedesign"><code>parsedesign</code></a>. </p><p><strong>Fields</strong></p><p><code>designtype::Symbol</code>: type of the subpopulation design. It must be :commoncross, :breedcross, :juncdist, or :matescheme. </p><p><code>founders::Union{Nothing,AbstractVector}</code>: founders for the subpopulation. </p><p><code>designcode::Union{Nothing,AbstractString}</code>: string code for the design. </p><p><code>pedigree::Union{Nothing, Pedigree}</code>: pedigree for the design </p><p><code>matescheme::Union{Nothing, MateScheme}</code>: mate schemes for the design</p><p><code>juncdist::Union{Nothing, JuncDist}</code>: junctdist for the design. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.MagicPed" href="#MagicBase.MagicPed"><code>MagicBase.MagicPed</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicPed</code></pre><p>mutable struct that stores pedigree information</p><pre><code class="nohighlight hljs">MagicPed(designinfo,founderinfo,offspringinfo)</code></pre><p>inner constructor. See also <a href="#MagicBase.readmagicped"><code>readmagicped</code></a>.</p><p><strong>Fields</strong></p><p><code>designinfo::Union{Nothing, Dict{String,DesignInfo},Pedigree}</code>: specifies population designinfo. A designinfo::Pedigree specifies the designinfo in pedigree. See also <a href="#Pedigrees.Pedigree"><code>Pedigree</code></a>.  A designinfo::Dict{String,DesignInfo} specifies designinfo for each subpopulation.  See also <a href="#MagicBase.DesignInfo"><code>DesignInfo</code></a>. </p><p><code>founderinfo::DataFrame</code>: founder information. </p><p><code>offspringinfo::DataFrame</code>: offspring information. The column names are [:individual,:member,:ishomozygous,:isfglexch,:gender]. The :ishomozygous column specifies if the individual is homozygous. If design is set by juncdist::JuncDist, the :member column is set to &quot;juncdist&quot;. If design is set by a string designcode::AbstractString, the :member column is set to the last non-male member of the pedigree. If design is set by designinfo::Pedigree, the :member column is associated to the :member column of the pedigree.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.formmagicped" href="#MagicBase.formmagicped"><code>MagicBase.formmagicped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">formmagicped(designinfo, popsize)</code></pre><p>form magicped::MagicPed form magicped for a non-divided pouplation of popsize. </p></div></section><section><div><pre><code class="language-julia hljs">formmagicped(pedigree, popsize)</code></pre><p>form magicped::MagicPed from pedigree and popsize. </p></div></section><section><div><pre><code class="language-julia hljs">formmagicped(designcode, popsize)</code></pre><p>form magicped::MagicPed from designcode and popsize. </p></div></section><section><div><pre><code class="language-julia hljs">formmagicped(juncdist, popsize)</code></pre><p>form magicped::MagicPed from juncdist and popsize. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.readmagicped" href="#MagicBase.readmagicped"><code>MagicBase.readmagicped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readmagicped(pedfile; commentstring=&quot;##&quot;,workdir=pwd())</code></pre><p>read a CSV formatted <code>pedfile</code> and return magicped::MagicPed. </p><p><strong>Positional arguments</strong></p><p><code>pedfile::AbstractString</code>: saves the pedigre information: designinfo and offspringinfo.  The designinfo can be provided in three formats: pedigree, mating design code,  and junction distribution. </p><p><strong>Keyword arguments</strong></p><p><code>commentstring::AbstractString=&quot;##&quot;</code>: the lines beginning with commentstring are ignored in pedfile.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading pedfile.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.savemagicped" href="#MagicBase.savemagicped"><code>MagicBase.savemagicped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">savemagicped(sink,magicped; delim=&#39;,&#39;,workdir=pwd())</code></pre><p>save magicped into sink. see [<code>readmagicped</code>] for reading saved output file.</p><p><strong>Positional arguments</strong></p><p><code>sink::Union{IO,AbstractString}</code>: output file or IO.</p><p><code>magicped::MagicPed</code>: a struct for storing pedidgree info.</p><p><strong>Keyword arguments</strong></p><p><code>delim::AbstractChar=&#39;,&#39;</code>: delimitor character.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading genfile and pedfile.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.plotmagicped" href="#MagicBase.plotmagicped"><code>MagicBase.plotmagicped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotmagicped(magicped; kwargs...)</code></pre><p>plot magicped. </p><p><strong>Keyword arguments</strong></p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred, and otherwise outbred.</p><p><code>outfile::Union{Nothing,AbstractString}=nothing</code>: if nothing, not save the plot, and otherwise save to outfile. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.MagicGeno" href="#MagicBase.MagicGeno"><code>MagicBase.MagicGeno</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicGeno</code></pre><p>mutable struct that stores genotypic data</p><pre><code class="nohighlight hljs">MagicGeno(magicped,markermap,foundergeno,offspringgeno,deletion,correction)</code></pre><p>inner constructor. See also <a href="#MagicBase.formmagicgeno"><code>formmagicgeno</code></a>.</p><p><strong>Fields</strong></p><p><code>magicped::MagicPed</code>: breeding pedigree information. See also <a href="#MagicBase.MagicPed"><code>MagicPed</code></a>.</p><p><code>markermap::Vector{DataFrame}</code>: marker map for each chromosome. <code>markermap[c]</code> gives the markermap of chromosome c.</p><p><code>foundergeno::Vector{Matrix}</code>: genotypic data in founders. <code>foundergeno[ch][m,i]</code> gives the genotype of founder i at marker m of chromosome ch.</p><p><code>offspringgeno::Vector{Matrix}</code>: genotypic data in offspring. <code>offspringgeno[ch][m,i]</code>  gives the genotype of offspring i at marker m of chomosome ch.</p><p><code>misc::Dict{String, DataFrame}</code>: contains information such as (1) &quot;deletion&quot; dataframe for markers that were removed from markermap, (2) &quot;correction&quot; dataframe for parental error correction.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.formmagicgeno" href="#MagicBase.formmagicgeno"><code>MagicBase.formmagicgeno</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">formmagicgeno(genofile, nfounder;
    isfounderinbred, formatpriority, isphysmap, recomrate, commentstring)</code></pre><p>form magicgeno::MagicGeno from the <code>genofile</code> and the number <code>nfounder</code> of founders. Assume that founders&#39; columns are on the left of offspring columns.</p></div></section><section><div><pre><code class="language-julia hljs">formmagicgeno(genofile, pedinfo;
    formatpriority, isphysmap, recomrate, commentstring)</code></pre><p>form magicgeno from the <code>genofile</code> and the pedigree information <code>pedinfo</code>.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code>: genotypic data file with extension &quot;.vcf&quot; or &quot;.vcf.gz&quot;.</p><p><code>pedinfo::AbstractString</code>: <code>designcode</code> or <code>pedfile</code></p><ul><li><p><code>pedfile</code>: See <a href="#MagicBase.readmagicped"><code>readmagicped</code></a> for the format of a pedigree file.</p></li><li><p><code>designcode</code>: a string designcode for a breeding population.  See <a href="#MagicBase.parsedesign"><code>parsedesign</code></a> for details.</p></li></ul><p><strong>Keyword arguments</strong></p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred, and otherwise outbred.</p><p><code>formatpriority::AbstractVector=[&quot;AD&quot;,&quot;GT&quot;]</code>: the priority of genotype  formats when parasing input vcf genofile.  </p><p><code>isphysmap::Bool=false</code>: if ture, transform physical map into genetic map using recomrate and overwrite the exist genetic map.  If false, keep input physical and/or genetic map.&quot;</p><p><code>recomrate::Real=1.0</code>, average recombation rate in cM per Mbp. Valid only if isphysmap = true.</p><p><code>commentstring::AbstractString=&quot;##&quot;</code>: the lines beginning with commentstring are ignored in genofile or pedfile.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading genofile and pedfile.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.savemagicgeno" href="#MagicBase.savemagicgeno"><code>MagicBase.savemagicgeno</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">savemagicgeno(outfile,magicgeno,workdir=pwd(),delim=&#39;,&#39;)</code></pre><p>save genotypic data of magicgeno::MagicGeno into outfile</p><p><strong>Positional arguments</strong></p><p><code>outfile::AbstractString</code>: filename for saving results.</p><p><code>magicgeno::MagicGeno</code>: a struct returned by <a href="#MagicBase.formmagicgeno"><code>formmagicgeno</code></a>.</p><p><strong>Keyward arguments</strong></p><p><code>missingstring::AbstractString = &quot;NA&quot;</code>: string for missing values <code>workdir::AbstractString = pwd()</code>: directory for writing outfile.</p><p><code>delim::AbstractChar=&#39;,&#39;</code>: delimitor character</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.savegenodata" href="#MagicBase.savegenodata"><code>MagicBase.savegenodata</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">savegenodata(outfile,magicgeno; missingstring=&quot;NA&quot;,workdir=pwd(),delim=&#39;,&#39;)</code></pre><p>save genotypic data of magicgeno::MagicGeno into outfile</p><p><strong>Positional arguments</strong></p><p><code>outfile::AbstractString</code>: output genofile for saving genotypic data.</p><p><code>magicgeno::MagicGeno</code>: a struct returned by <a href="#MagicBase.formmagicgeno"><code>formmagicgeno</code></a>.</p><p><strong>Keyward arguments</strong></p><p><code>workdir::AbstractString = pwd()</code>: directory for writing outfile.</p><p><code>delim::AbstractChar=&#39;,&#39;</code>: delimitor character</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.MagicAncestry" href="#MagicBase.MagicAncestry"><code>MagicBase.MagicAncestry</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicAncestry</code></pre><p>mutable struct that stores the results of haplotype reconstruction by magicreconstruct. </p><pre><code class="nohighlight hljs">MagicAncestry(magicped,markermap,foundergeno,statespace,
    viterbipath, diploprob,genoprob,haploprob,loglike,misc)</code></pre><p>inner constructor. See also <a href="#MagicBase.readmagicancestry"><code>readmagicancestry</code></a>.</p><p><strong>Fields</strong></p><p><code>magicped::MagicPed</code>: breeding pedigree information. See also <a href="#MagicBase.MagicPed"><code>MagicPed</code></a>.</p><p><code>markermap::Vector{DataFrame}</code>: marker map for each chromosome. <code>markermap[c]</code> gives the marker map of chromosome c. </p><p><code>foundergeno::Vector{Matrix}</code>: haplotypes in founders. <code>foundergeno[c][m,f]</code> gives the founder f at marker m of chromosome c.</p><p><code>statespace::AbstractDict</code>: definition of ancestral haplotype states corresponding to <code>haploprob</code>.</p><p><code>viterbipath::Union{Nothing,Vector{Matrix}}</code>: optimal state path obtained by the Viterbi algorithm. <code>viterbipath[c][m,o]</code> gives the ancestral diplotype state of offspring o at marker m for chromosome c.</p><p><code>diploprob::Union{Nothing,Vector{Vector{SparseArrays.SparseMatrixCSC}}}</code>: marginal posterior probabilities for diplotypes. <code>diploprob[c][o][m,s]</code> gives the probability of offspring o at ancestral diplotype state s for marker m of chromosome c.</p><p><code>genoprob::Union{Nothing,Vector{Vector{SparseArrays.SparseMatrixCSC}}}</code>: marginal posterior probabilities for genotypes. <code>genoprob[c][o][m,s]</code> gives the probability of offspring o at ancestral genotype state s for marker m of chromosome c.</p><p><code>haploprob::Union{Nothing,Vector{Vector{SparseArrays.SparseMatrixCSC}}}</code>: marginal posterior probabilities for haplotypes. <code>genoprob[c][o][m,s]</code> gives the probability of offspring o at ancestral haplotype state s for marker m of chromosome c.</p><p><code>inbredcoef::Union{Nothing,Vector{Matrix}}</code>: realized inbreeding coefficients.  <code>inbredcoef[c][m,o]</code> gives the inbreeding coefficients at marker m of chromosome c in offpsring o. </p><p><code>loglike::AbstractMatrix</code>: loglike[c,o] gives the log likelihood for chromsome c of offspring o.</p><p><code>misc::Dict{String, DataFrame}</code>: contains information such as (1) &quot;deletion&quot; dataframe for markers that were removed from markermap, (2) &quot;correction&quot; dataframe for parental error correction.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.readmagicancestry" href="#MagicBase.readmagicancestry"><code>MagicBase.readmagicancestry</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">readmagicancestry(ancestryfile, workdir=pwd(),tempdirectory=tempdir())</code></pre><p>read ancestryfile in the directory workdir and return magicancestry::MagicAncestry. See <a href="#MagicBase.MagicAncestry"><code>MagicAncestry</code></a>. </p><p><strong>Positional argument</strong></p><p><code>ancestryfile</code>: file storing magicancestry that is generated by <a href="#MagicBase.savemagicancestry"><code>savemagicancestry</code></a>. It results from <code>magicreconstruct</code>.</p><p><strong>Keyward arguments</strong></p><p><code>workdir::AbstractString = pwd()</code>: directory for reading ancestryfile,</p><p><code>tempdirectory::AbstractString=tempdir()</code>: temparary directory. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.savemagicancestry" href="#MagicBase.savemagicancestry"><code>MagicBase.savemagicancestry</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">savemagicancestry(outputfile,magicancestry,workdir=pwd())</code></pre><p>save magicancestry into outputfile.</p><p><strong>Positional arguments</strong></p><p><code>outputfile::AbstractString</code>: file for saving magicancestry. </p><p><code>magicancestry::MagicAncestry</code>: a struct returned by magicreconstruct. </p><p><strong>Keyward arguments</strong></p><p><code>workdir::AbstractString = pwd()</code>: directory for writing outputfile.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.parsedesign" href="#MagicBase.parsedesign"><code>MagicBase.parsedesign</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">parsedesign(designcode; kwargs...)</code></pre><p>parse string designcode into <a href="#MagicBase.DesignInfo"><code>DesignInfo</code></a></p><p><strong>Keyword Argument</strong></p><p><code>founders::Union{Nothing,AbstractVector}=nothing</code>: a list of founders. </p><p><code>popid=&quot;pop&quot;</code>: population id. </p><p><code>fixed_nself::Integer=20</code>: number of selfing generation for designcode in form of pedcode=&gt;FIXED. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.parsebreedped" href="#MagicBase.parsebreedped"><code>MagicBase.parsebreedped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">parsebreedped(pedfile; fixed_nself=10, outfile, commentstring=&quot;##&quot;,workdir=pwd())</code></pre><p>convert a breedped pedfile into the magicped outfile. The first 3 columns of breedpedfile must be sample, pedcode, nself.</p><p><strong>Keyword arguments</strong></p><p><code>fixed_nself::Integer = 20</code>: interprete &quot;FIXED&quot; in nself column as 20. </p><p><code>delim::AbstractChar=&#39;,&#39;</code>: text delimitor of input pedfile. </p><p><code>commentstring::AbstractString=&quot;##&quot;</code>: the lines beginning with commentstring are ignored in pedfile.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading pedfile.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.resetmap" href="#MagicBase.resetmap"><code>MagicBase.resetmap</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">resetmap(vcffile,mapfile;
    missingstring, commentstring, outstem, workdir)</code></pre><p>exports a new vcf file with marker map replaced with mapfile. </p><p><strong>Positional arguments</strong></p><p><code>vcffile::AbstractString</code>: genotypic data file with extension &quot;.vcf&quot; or &quot;.vcf.gz&quot;.</p><p><code>mapfile::AbstractString</code>: file for marker map, it can either be in VCF format or in CSV format.   For CSV-format, it must contain at least five columns: marker, linkagegroup, poscm, physchrom, physposbp.   The values are represented by missingstring. </p><p><strong>Keyword arguments</strong></p><p><code>missingstring::AbstractString=&quot;NA&quot;</code>: string representing missing value. </p><p><code>commentstring::AbstractString=&quot;##&quot;</code>: the lines beginning with commentstring are ignored in vcffile or mapfile.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code>: stem of output filenames.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading genfile and pedfile.    </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.rabbitped_mma2jl" href="#MagicBase.rabbitped_mma2jl"><code>MagicBase.rabbitped_mma2jl</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">rabbitped_mma2jl(mmapedfile; kwargs...)</code></pre><p>convert input pedfile of Mathematica-version RABBIT into pedfile for Julia-version RABBIT.</p><p><strong>keyword arguments</strong></p><p><code>ishomozygous::Bool = false</code>: if true, offspring is homozygous. </p><p><code>isfglexch::Bool = false</code>: if true, offspring is produced with rand parent ordering.</p><p><code>workdir::AbstractString=pwd()</code>: working directory for input and output files.</p><p><code>outfile::AbstractString = &quot;outstem_ped.csv&quot;</code>: output filename.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.rabbitgeno_mma2jl" href="#MagicBase.rabbitgeno_mma2jl"><code>MagicBase.rabbitgeno_mma2jl</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">rabbitgeno_mma2jl(mmagenofile; kwargs...)</code></pre><p>convert input genofile of Mathematica-version RABBIT into genofile for Julia-version RABBIT. </p><p><strong>keyword arguments</strong></p><p><code>workdir::AbstractString=pwd()</code>: working directory for input and output files.</p><p><code>outfile::AbstractString = &quot;outstem_geno.vcf.gz&quot;</code>: output filename.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.arrayfile2vcf" href="#MagicBase.arrayfile2vcf"><code>MagicBase.arrayfile2vcf</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">arrayfile2vcf(arrayfile; keyargs...)</code></pre><p>extract pedfile from arrayfile. Work only for a non-subdivided population.  </p><p><strong>Positional arguments</strong></p><p><code>arrayfile::AbstractString</code>: SNP array genofile. </p><p><strong>Keyword arguments</strong></p><p><code>delmultiallelic::Bool = true,</code>: if true, delete markers with #alleles &gt;= 3. </p><p><code>delim= &quot;,&quot;</code>: text delimiter. </p><p><code>outstem::AbstractString=&quot;outstem&quot;</code>: stem of output filename.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= outstem*&quot;_arrayfile2vcf.log&quot;</code>: log file or IO for writing log. If it is nothing, no log file.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.hapmap2vcf" href="#MagicBase.hapmap2vcf"><code>MagicBase.hapmap2vcf</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">hapmap2vcf(hapmapfile; keyargs...)</code></pre><p>convert from a hapmap genofile into a vcf genofile. </p><p><strong>Positional arguments</strong></p><p><code>hapmapfile::AbstractString</code>: hapmap genofile. </p><p><strong>Keyword arguments</strong></p><p><code>delmultiallelic::Bool = true,</code>: if true, delete markers with #alleles &gt;= 3. </p><p><code>delim= &quot;,&quot;</code>: text delimiter. </p><p><code>missingallele::AbstractString=&quot;-&quot;</code>: string for missing allele</p><p><code>outstem::AbstractString=&quot;outstem&quot;</code>: stem of output filename.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= outstem*&quot;_arrayfile2vcf.log&quot;</code>: log file or IO for writing log. If it is nothing, no log file.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.merge_vcffiles" href="#MagicBase.merge_vcffiles"><code>MagicBase.merge_vcffiles</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">merge_vcffiles(vcffiles; outstem, workdir)</code></pre><p>merge vcffiles into a single vcf genofile. </p><p><strong>Positional arguments</strong></p><p><code>vcffiles::AbstractVector</code>: a list of vcf genofile. </p><p><strong>Keyword arguments</strong></p><p><code>outstem::AbstractString=&quot;outstem&quot;</code>: stem of output filename.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.merge_arrayfiles" href="#MagicBase.merge_arrayfiles"><code>MagicBase.merge_arrayfiles</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">merge_arrayfiles(arrayfiles; outstem, workdir)</code></pre><p>merge arrayfiles into a single SNP array genofile. </p><p><strong>Positional arguments</strong></p><p><code>arrayfiles::AbstractVector</code>: a list of SNP array genofile. </p><p><strong>Keyword arguments</strong></p><p><code>missingallele::AbstractString = &quot;-&quot;</code>: string for missing allele. </p><p><code>outstem::AbstractString=&quot;outstem&quot;</code>: stem of output filename.</p><p><code>outext::AbstractString=&quot;.csv.gz&quot;</code>: extension of output file.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= outstem*&quot;_merge_arrayfiles.log&quot;</code>: log file or IO for writing log. If it is nothing, no log file.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.merge_pedfiles" href="#MagicBase.merge_pedfiles"><code>MagicBase.merge_pedfiles</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">merge_pedfiles(pedfiles; keyargs...)</code></pre><p>merge pedfiles into a single pedfile. </p><p><strong>Positional arguments</strong></p><p><code>pedfiles::AbstractString</code>: a list of pedfiles. </p><p><strong>Keyword arguments</strong></p><p><code>isped::Bool</code>: if true, designinfo in pedfiles is in form of pedigree, and otherwise designcode. </p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred. </p><p><code>outstem::AbstractString=popid</code>: stem of output filename.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.array_extract_pedfile" href="#MagicBase.array_extract_pedfile"><code>MagicBase.array_extract_pedfile</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">array_extract_pedfile(arrayfile; keyargs...)</code></pre><p>extract pedfile from arrayfile. Work only for a non-subdivided population.  </p><p><strong>Positional arguments</strong></p><p><code>arrayfile::AbstractString</code>: SNP array genofile. </p><p><strong>Keyword arguments</strong></p><p><code>designcode::AbstractString</code>: designcode</p><p><code>popid::AbstractString</code>: population id. </p><p><code>delim= &quot;,&quot;</code>: text delimiter. </p><p><code>outstem::AbstractString=popid</code>: stem of output filename.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.vcf_pad_samples" href="#MagicBase.vcf_pad_samples"><code>MagicBase.vcf_pad_samples</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">vcf_pad_samples(vcffile; keyargs...)</code></pre><p>pad samples into vcffiles with all genotypes being missing. </p><p><strong>Positional arguments</strong></p><p><code>vcffile::AbstractString</code>: vcf genofile. </p><p><strong>Keyword arguments</strong></p><p><code>padsamples::AbstractVector</code>: a list of samples to be padded </p><p><code>commentstring::AbstractString=&quot;##&quot;</code>: the lines beginning with  are ignored</p><p><code>outstem::AbstractString=popid</code>: stem of output filename.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.vcf_extract_pedfile" href="#MagicBase.vcf_extract_pedfile"><code>MagicBase.vcf_extract_pedfile</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">vcf_extract_pedfile(vcffile; keyargs...)</code></pre><p>extract pedfile from vcffile. Work only for a non-subdivided population.  </p><p><strong>Positional arguments</strong></p><p><code>vcffile::AbstractString</code>: vcf genofile. </p><p><strong>Keyword arguments</strong></p><p><code>designcode::AbstractString</code>: designcode</p><p><code>ishomozygous::Bool=false</code>: specify if offspring are completely homozygous.</p><p><code>isfglexch::Bool=true</code>: specify if founders are exchangeable.</p><p><code>popid::AbstractString=&quot;pop&quot;</code>: population id. </p><p><code>outstem::AbstractString=&quot;outstem&quot;</code>: stem of output filename.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.vcffilter" href="#MagicBase.vcffilter"><code>MagicBase.vcffilter</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">vcffilter(vcffile; keyargs...)</code></pre><p>filter markers line by line. </p><p><strong>Positional arguments</strong></p><p><code>vcffile::AbstractString</code>: vcf genofile. </p><p><strong>Keyword arguments</strong></p><p><code>setmarkerid::Union{Nothing,Bool}=nothing</code>: if true, set markerid. If it is nothing, setmarkerid = true only if markerid is missing</p><p><code>delsamples::Union{Nothing,AbstractVector}=nothing</code>: list of sample IDs to be deleted. If it is nothing, no deletion of samples </p><p><code>deldupe::Bool=false</code>: if true, delete sucessive markers that have exactly duplicated genotypes in format of GT</p><p><code>delmultiallelic::Bool=true</code>: if true, delete markers with &gt;2 alleles. </p><p><code>delmonomorphic::Bool=true</code>: if true, delete markers with single allele. </p><p><code>seqstretch::Integer=0</code>: delete non-initial markers in a sequence stretch of length &lt;= seqstretch (in bp), assuming marker are ordered by physical positions. If it is not positive, no filtering for short streches.</p><p><code>snp_maxmiss::Real = 0.99</code>: delete markers with missing fraction &gt; snp_maxmiss</p><p><code>snp_minmaf::Real = 0.01</code>: delete markers with minor allele frequency &lt; snp_minmaf</p><p><code>commentstring::AbstractString=&quot;##&quot;</code>: the lines beginning with  are ignored</p><p><code>outstem::AbstractString=popid</code>: stem of output filename.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading and saving files.</p><p><code>logfile::AbstractString = outstem*&quot;_vcffilter.log&quot;</code>: log filename. </p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.pedfile_designcode2ped" href="#MagicBase.pedfile_designcode2ped"><code>MagicBase.pedfile_designcode2ped</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">pedfile_designcode2ped(pedfile; commentstring=&#39;#&#39;,workdir=pwd())</code></pre><p>convert a pedfile from <code>designinfo</code> being <code>designcode</code> to <code>pedigree</code>.</p><p><strong>Keyword arguments</strong></p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred, and otherwise outbred. </p><p><code>commentstring::AbstractString=&quot;##&quot;</code>: the lines beginning with commentstring are ignored in pedfile.</p><p><code>workdir::AbstractString=pwd()</code>: directory for reading pedfile.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.plotcondprob" href="#MagicBase.plotcondprob"><code>MagicBase.plotcondprob</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotcondprob(magicancestry,offspring=nothing,probtype=&quot;haploprob&quot;,
    colorgradient = cgrad([:white,:blue,:red]),,
    boundaryline = (1.5, :gray),
    truemarker=(:star, 5, 0.5,:gray,stroke(:gray)),
    truefgl=nothing,
    outfile::Union{Nothing, AbstractString}=nothing,
    plotkeyargs...)</code></pre><p>plot heatmap for conditional probability.</p><p><strong>Positional arguments</strong></p><p><code>magicancestry::MagicAncestry</code>: magicancestry returned from <code>magicreconstruct</code>.</p><p><strong>Keyword arguments</strong></p><p><code>probtype::AbstractString=&quot;haploprob&quot;</code>: specify type of condprob</p><p><code>offspring::Union{Nothing,Integer}=nothing</code>: offsprign index. By default, a random offspring index. </p><p><code>colorgradient::ColorGradient=cgrad([:white,:blue,:red]),</code>: color gradient for heatmap</p><p><code>boundaryline=(1.5,:gray)</code>: vertical lines for chromosome boundaries.</p><p><code>truemarker=(:star, 5, 0.5,:gray,stroke(:gray))</code>: scatter markers for true ancestral states.</p><p><code>truefgl::Union{Nothing,MagicGeno}=nothing</code>: provides true ancestral origins. </p><p><code>outfile::Union{Nothing, AbstractString}=nothing</code>: if nothing, not save the plot, and otherwise save to outfile.  </p><p><code>plotkeyargs...</code>: other Plots.plot keyward arguments. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.animcondprob" href="#MagicBase.animcondprob"><code>MagicBase.animcondprob</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">animcondprob(magicancestry,fps=1,outstem=&quot;&quot;,kewargs...)</code></pre><p>animation for plots of conditional probability.</p><p><strong>Positional arguments</strong></p><p><code>magicancestry::MagicAncestry</code>: magicancestry returned from <code>magicreconstruct</code>.</p><p><strong>Keyword arguments</strong></p><p><code>fps::Real=1</code>: number of frames per seconds.</p><p><code>outfile::AbstractString=&quot;condprob.gif&quot;</code>: output file for saving animation.</p><p>see <a href="#MagicBase.plotcondprob"><code>plotcondprob</code></a> for keyargs.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.plotmarkermap" href="#MagicBase.plotmarkermap"><code>MagicBase.plotmarkermap</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotmarkermap(mapx, mapy;
    boundaryline = (1.5,:dot,:black),
    markersize = 1.5,
    isannotate= true,
    maplabels=[&quot;mapx(cM)&quot;, &quot;mapy(cM)&quot;],
    isphysmap = [false,false],
    plotkeyargs...
)</code></pre><p>plot postions of mapx vs those of mapy.</p><p><strong>Positional arguments</strong></p><p><code>mapx::Vector{DataFrame}</code>: marker map for all chromosomes.</p><p><code>mapy::Vector{DataFrame}</code>: comparing map for all chromosomes.</p><p><strong>Keyword arguments</strong></p><p><code>boundaryline = (1.0,:dot,:gray)</code>: vertical lines for chromosome boundaries.</p><p><code>markersize::Real= size(mapx,1) &lt;= 1000 ? 3.0 : 1.5</code>: size of scatter markers.</p><p><code>isannotate::Bool=true</code>: if ture,  annotate chromosome ID and kendall correlation.</p><p><code>mmaplabels::Union{Nothing,AbstractString}=nothing</code>: labels of comparing marker maps.</p><p><code>isphysmap::AbstractVector = falses(2)</code>: specify if mapx and/or map are physical maps. </p><p><code>plotkeyargs...</code>: other Plots.plot keyward arguments. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.plotrecombreak" href="#MagicBase.plotrecombreak"><code>MagicBase.plotrecombreak</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">plotrecombreak(magicancestry,chr=1,
    colorgradient = ColorGradient([:yellow,:blue,:red]),
    truefgl=nothing)</code></pre><p>plot recombination breakpoints.</p><p><strong>Positional arguments</strong></p><p><code>magicancestry::MagicAncestry</code>: magicancestry returning from <code>magicreconstruct</code>.</p><p><strong>Keyword arguments</strong></p><p><code>chr::Integer=1</code>: chromosome index.</p><p><code>colorgradient::ColorGradient=cgrad([:white,:blue,:red])</code>: color gradient for heatmap</p><p><code>truefgl::Union{Nothing,MagicGeno}</code>: provides true ancestral origins. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicSimulate" href="#MagicSimulate"><code>MagicSimulate</code></a> — <span class="docstring-category">Module</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicSimulate</code></pre><p>a package for simulating genotyping data in multiparental populations.  See also <a href="#MagicSimulate.magicsimulate"><code>magicsimulate</code></a>.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicSimulate.simfhaplo" href="#MagicSimulate.simfhaplo"><code>MagicSimulate.simfhaplo</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">simfhaplo(; kwargs...)</code></pre><p>simulate founder haplotypes. </p><p><strong>Keyword arguments</strong></p><p><code>nsnp</code>: number of markers.</p><p><code>nparent</code>: number of parents.</p><p><code>missingstring=&quot;NA&quot;</code>: string representing missing value. </p><p><code>chrlen::AbstractVector = 100 * ones(5)</code>:  specify genetic length of each chromosome. </p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred, and otherwise outbred.</p><p><code>outfile = &quot;sim_fhaplo.vcf.gz&quot;</code>: output filename. </p><p><code>workdir::AbstractString = pwd()</code> specifies the working directory.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicSimulate.magicsimulate" href="#MagicSimulate.magicsimulate"><code>MagicSimulate.magicsimulate</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicsimulate(fhaplofile, pedinfo; kwargs...)</code></pre><p>simulate genotypic data from founder haplotypes in <code>fhaplofile</code> and pedigree information in <code>pedinfo</code>.</p><p><strong>Positional arguments</strong></p><p><code>fhaplofile::AbstractString</code> specifies file for founder haplotypes including marker map. The fhaplofile extension must be &quot;.vcf&quot; or &quot;.csv&quot;.</p><p><code>pedinfo::Union{AbstractString,MateScheme}</code> specifies pedigree information via a matescheme (e.g. MateScheme(8,[&quot;Pairing&quot;,&quot;Selfing&quot;],[3,6])),  or via a pedigree file (if pedinfo[end-3:end]==&quot;.csv&quot;),  or via a string designcode (e.g. &quot;8ril-self6&quot;).</p><p><strong>Keyword arguments</strong></p><p><code>isfounderinbred::Bool=true</code> if true, the founders are inbred. For inbred founders, the heterozygous genotypes are set to missing. For outbred founders, genotypes must be phased.</p><p><code>popsize::Union{Nothing,Integer}=200</code> specifies the population size, valid only if the <code>pedinfo</code> is specified via a designcode or matescheme.</p><p><code>error_randallele::Union{Nothing,Real}=0.0</code> specifices genotyping error model.  The error model follows uniform allele model with probability error<em>randallele,  and it follows uniform genotype model with probability 1-error</em>randallele. If it is nothing, error_randallele  is given by the interally estimated non-ibd probability. </p><p><code>foundererror::Distribution=Beta(1,199)</code> specifies that the probability distribution of genotyping error rate at a marker in founders. At a given marker, founders have the same error rate. </p><p><code>offspringerror::Distribution=Beta(1,199)</code> specifies that the probability distribution of genotyping error rate at a marker in offspring. At a given marker, offspring have the same error rate. </p><p><code>foundermiss::Distribution=Beta(1,9)</code> specifies that the probability distribution of the fraction of missing genotypes at a marker in founders.</p><p><code>offspringmiss::Distribution=Beta(1,9)</code> specifies that the probability distribution of the fraction of missing genotypes at a marker in offspring.</p><p><code>seqfrac::Real=0.0</code> specifies the fraction of markers being genotyped by sequencing; the rest markers are genotyped by SNP array.</p><p><code>seqerror::Distribution = Beta(1,199)</code> specifies the probability diestribution of sequencing error rate among markers. </p><p><code>allelebalancemean::Distribution = Beta(10,10)</code> specifies the probability distribution of the mean sequencing allelic balance among markers. </p><p><code>allelebalancedisperse::Distribution = Exponential(0.05)</code> specifies overdispersion parameter for the probability distribution of allelebalancemean among individuals at a marker. </p><p><code>seqdepth::Distribution = Gamma(2, 5)</code> specifies the probability distribution of the mean read depth at a marker. </p><p><code>seqdepth_overdispersion::Distribution = Gamma(1,1)</code> specifies the probability distribution of over-dispersion of read depths among individuals at a marker.  Given the mean depth lam at a marker, the read depth of an individual follows a NegativeBionomial(r,p), such that mean lam = r(1-p)/p, and variance = r(1-p)/p^2 = r(1+lam/r) = r(1+seqdepth<em>overdispersion) where seqdepth</em>overdispersion = lam/r. seqdepth_overdispersion = 0 denotes no over-dispersion. </p><p><code>isobligate::Bool=false</code> specifies whether there must be at least one crossover event</p><p><code>interference::Integer=0</code> specifies the strength of chiasma interference. By default, no chiasma interference. The recombination breakpoins are obtained by taking every (1+interference) points that follow a Poisson distribution along chromosome.</p><p><code>ispheno::Bool=false</code> specifies whether to simulate phenotypes.</p><p><code>pheno_nqtl::Integer=1</code> specifies the number of QTLs in simulating phenotypes.</p><p><code>pheno_h2::Real= 0.5</code> specifies heritablity in simulating phenotypes.</p><p><code>select_nqtl::Integer=1</code> specifies the number of QTLs in simulating trait for artifical selection.</p><p><code>select_prop::Real = 1.0</code> specifies the proportion of zygotes selected in artifical selection. By default, no artifical selection.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code> specifies the stem of output filenames.</p><p><code>workdir::AbstractString = pwd()</code> specifies the working directory.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicsimulate(&quot;fhaplo.vcf.gz&quot;,&quot;ped.csv&quot;)
julia&gt; magicsimulate(&quot;fhaplo.vcf.gz&quot;,&quot;8ril-self6&quot;; popsize=800)</code></pre></div></section><section><div><pre><code class="language-julia hljs">magicsimulate(pedinfo; kwargs...)</code></pre><p>simulates ancestral blocks from pedinfo. </p><p><strong>Positional arguments</strong></p><p><code>pedinfo::Union{AbstractString,MateScheme}</code> specifies pedigree information via a matescheme (e.g. MateScheme(8,[&quot;Pairing&quot;,&quot;Selfing&quot;],[3,6])),  or via a pedigree file (if pedinfo[end-3:end]==&quot;.csv&quot;),  or via a string designcode (e.g. &quot;8ril-self6&quot;).</p><p><strong>Keyword arguments</strong></p><p><code>isfounderinbred::Bool=true</code> if true, the founders are inbred. For inbred founders, the heterozygous genotypes are set to missing. For outbred founders, genotypes must be phased.</p><p><code>popsize::Union{Nothing,Integer}=200</code> specifies the population size, valid only if the <code>pedinfo</code> is specified via a designcode or matescheme.</p><p><code>chrlen::AbstractVector= 100*ones(5)</code> specifies lengths (cM) for each chromosome.</p><p><code>isobligate::Bool=false</code> specifies whether there must be at least one crossover event</p><p><code>interference::Integer=0</code> specifies the strength of chiasma interference. By default, no chiasma interference. The recombination breakpoins are obtained by taking every (1+interference) points that follow a Poisson distribution along chromosome.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code> specifies the stem of output filenames.</p><p><code>workdir::AbstractString = pwd()</code> specifies the working directory.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p><p><strong>Outputs</strong></p><table><tr><th style="text-align: left">Output file</th><th style="text-align: left">Description</th></tr><tr><td style="text-align: left"><code>outstem_ped.csv</code></td><td style="text-align: left">simulated pedigree file</td></tr><tr><td style="text-align: left"><code>outstem_truecontfgl.csv</code></td><td style="text-align: left">truevalues of continuous origin-genotypes</td></tr></table><p>Here fgl denotes founder genome labels, and origin-genotypes denote genotypes with each fgl  being regarded as a distinct allele.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicsimulate(&quot;8ril-self6&quot;; popsize=800)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicFilter" href="#MagicFilter"><code>MagicFilter</code></a> — <span class="docstring-category">Module</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicFilter</code></pre><p>a package for filtering markers and individuals in connected multiparental populations.  Export one function:  <a href="#MagicFilter.magicfilter"><code>magicfilter</code></a>.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicFilter.magicfilter" href="#MagicFilter.magicfilter"><code>MagicFilter.magicfilter</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicfilter(genofile, pedinfo;
    formatpriority, isphysmap, recomrate, commentstring,kwargs...)</code></pre><p>filter mrkers and founders/offspring from genofile and pedinfo.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code> genotypic data file.</p><p><code>pedinfo::Union{MagicBase.JuncDist,AbstractString}</code> specifies pedigree information via a pedigree fille or a string designcode or via a struct juncdist::JuncDist.</p><p><strong>Keyword arguments</strong></p><p>See <a href="#MagicBase.formmagicgeno"><code>formmagicgeno</code></a> for the arguments (<code>formatpriority</code>, <code>isphysmap</code>, <code>recomrate</code>,<code>commentstring</code>) that are used for formming magicgeno.  Note that formatpriority=[&quot;AD&quot;,&quot;GT&quot;] by default. </p><p>See <a href="#MagicFilter.magicfilter!"><code>magicfilter!</code></a> for kwargs.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicfilter(&quot;geno.vcf.gz&quot;,&quot;4ril_self3&quot;)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicFilter.magicfilter!" href="#MagicFilter.magicfilter!"><code>MagicFilter.magicfilter!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicfilter!(magicgeno::MagicGeno; kwargs...)</code></pre><p>removes bad mrkers and founders/offspring from magicgeno.</p><p><strong>Keyword arguments</strong></p><p><code>model::AbstractString=&quot;jointmodel&quot;</code>: prior dependence of ancestral prior process along the two homologous chromosomes within an offspring. It must be &quot;depmodel&quot;, &quot;indepmodel&quot;, or &quot;jointmodel&quot;. </p><p><code>likeparameters::LikeParameters=LikeParameters()</code>: specifies default genotyping error rates. </p><p><code>isfounderinbred::Bool=true</code>: specifies if fouonders are inbred</p><p><code>chrsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of chromosome indices.   <code>nothing</code> denotes all chromosomes. Delete chromosome indices that are out of range.</p><p><code>snpsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of marker indices within each chromosome.   <code>nothing</code> denotes all markers. Marker indices that are larger than the number of markers   within the chromosome are deleted.</p><p><code>threshcall::Real = model == &quot;depmodel&quot; ? 0.95 : 0.9</code>: threshold for genotype calling. The filtering is based on called genotypes. </p><p><code>min_subpop::Integer = 1</code>: delete subpopulations with size &lt; min_subpop.</p><p><code>min_nprogeny::Integer = 1</code>: delete founder and their progeny if the number of progeny &lt; min_nprogeny.</p><p><code>snp_monosubpop::Integer = 20</code>: a subpopulation is tested for monomorphic at a marker only if the number of observed genotypes &gt;= snp_mono_subpop.</p><p><code>snp_mono2miss::Union{Nothing,Bool} = true</code>: if true, all offspring genotypes in a monomorphic subpopulation are set to missing,  and otherwise only inconsistent offspring genotypes are corrected. And if nothing, offspring genotypes are not changed.</p><p><code>del_inconsistent::Bool = false</code>: if true, delete markers with inconsistent changes of founder genotypes. </p><p><code>snp_minmaf::Real = 0.05</code>: keep only markers if maf &gt;= snp_min_maf; maf denotes minor allele frequency.</p><p><code>snp_missfilter::Function=(fmiss,omiss)-&gt; omiss &lt;= 1.0 || fmiss &lt; 0.0</code>: keep only markers if snp_missfilter(fmiss, omiss); fmiss denotes missing fraction in founders, and omiss for offspring.</p><p><code>offspring_maxmiss::Real = 0.99</code>: delete offspring if its missing &gt; offspring_max_miss</p><p><code>isfilterdupe::Bool=false</code>: if true, remove duplicated offspring by their correlations.</p><p><code>offspring_maxcorr::Real = 0.99</code>: two offspring are duplciated if their correlation &gt;= offspring_maxcorr,</p><p><code>offspring_cutcorr::Real = 0.4</code>: pairwise offspring correlations that &lt; offspring_cutcorr are set to zeros. </p><p><code>isparallel::Bool=true</code>: if true, pefrom parallel multicore computing.</p><p><code>workdir::AbstractString=pwd()</code>: working directory for input and output files.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code> specifies the stem of output files.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= (isnothing(outstem) ? nothing : string(outstem,&quot;_magicreconstruct.log&quot;))</code>: log file or IO for writing log. If it is nothing, no log file.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicgeno = formmagicgeno(&quot;geno.vcf.gz&quot;,&quot;ped.csv&quot;)
julia&gt; magicfilter!(magicgeno)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicCall" href="#MagicCall"><code>MagicCall</code></a> — <span class="docstring-category">Module</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicCall</code></pre><p>a package for single site genotyping calling from sequence data in connected multiparental populations.  Export one function: <a href="#MagicCall.magiccall"><code>magiccall</code></a>.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicCall.magiccall" href="#MagicCall.magiccall"><code>MagicCall.magiccall</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magiccall(genofile, pedinfo; kwargs...)</code></pre><p>single marker genotype call from genofile and pedinfo.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code> genotypic data file.</p><p><code>pedinfo::Union{MagicBase.JuncDist,AbstractString}</code> specifies pedigree information via a pedigree fille or a string designcode or via a struct juncdist::JuncDist.</p><p><strong>Keyword arguments</strong></p><p><code>model::Union{AbstractString,AbstractVector}=&quot;jointmodel&quot;</code>:  prior depedence of ancestral prior process   along the two homologous chromosomes within an offspring. It must be &quot;depmodel&quot;,   &quot;indepmodel&quot;, or &quot;jointmodel&quot;. </p><p><code>likeparameters::LikeParameters=LikeParameters(peroffspringerror=0.0)</code>: parameters for genotypic data model.    If isinfererror = true, parameters with values being nothing will be inferred. </p><p><code>threshlikeparameters::ThreshLikeParameters=ThreshLikeParameters()</code>: markers with inferred likeparameters values &gt; threshlikeparameters values will be deleted. </p><p><code>priorlikeparameters::PriorLikeParameters=PriorLikeParameters(offspringerror=Beta(1.05,9),seqerror=Beta(1.05,9))</code>: priors for likelihood parameters</p><p><code>israndallele::Bool=true</code>: if true, genotyping error model follows the random allelic model, and otherwise the random genotypic model. </p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred, and otherwise they are outbred.</p><p><code>threshcall::Real = 0.9</code>: offspring genotypes are call if    the maximum posterior probability &gt; threshcall.</p><p><code>delmultiallelic::Bool=true</code>: if true, delete markers with &gt;=3 alleles. </p><p><code>delmonomorphic::Bool=true</code>: if true, delete monomorphic markers. </p><p><code>snp_minmaf::Real = 0.05</code>: delete makrs with minor allele frequency (MAF) &lt; 0.05. </p><p><code>snp_maxmiss::Real = 0.99</code>: delete makrs with genotype missing frequency &gt; 0.99. </p><p><code>israwcall::Bool= false</code>: if true, perform raw genotype calling. </p><p><code>isinfererror::Bool = !israwcall</code>: if true, infer marker specific likelihood parameters that have values of nothing in likeparameters. </p><p><code>isparallel::Bool=true</code>: if true, parallel multicore computing over chromosomes.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code>: stem of output filenames.</p><p><code>outext::AbstractString=&quot;.vcf.gz&quot;</code>: extension of output file for imputed geno.</p><p><code>logfile::Union{Nothing, AbstractString,IO} = outstem*&quot;_magiccall.log&quot;</code>: log file or IO for writing log. If it is nothing, no log file.</p><p><code>workdir::AbstractString=pwd()</code>: working directory for input and output files.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.  </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicMap" href="#MagicMap"><code>MagicMap</code></a> — <span class="docstring-category">Module</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicMap</code></pre><p>a package for genetic map construction in connected multiparental populations.  Export one function:  <a href="#MagicMap.magicmap"><code>magicmap</code></a>.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicMap.magicmap" href="#MagicMap.magicmap"><code>MagicMap.magicmap</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicmap(genofile, pedinfo; kwargs...)</code></pre><p>genetic map construction from genofile and pedinfo.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code> genotypic data file.</p><p><code>pedinfo::Union{MagicBase.JuncDist,AbstractString}</code> specifies pedigree information via a pedigree fille or a string designcode or via a struct juncdist::JuncDist.</p><p><strong>Keyword arguments</strong></p><p><code>formatpriority::AbstractVector=[&quot;GT&quot;,&quot;AD&quot;]</code>: the priority of genotype    formats when parasing input vcf genofile.  </p><p><code>model::AbstractString=&quot;jointmodel&quot;</code>:  prior depedence of ancestral prior process   along the two homologous chromosomes within an offspring. It must be &quot;depmodel&quot;,   &quot;indepmodel&quot;, or &quot;jointmodel&quot;. </p><p><code>likeparameters::LikeParameters=LikeParameters()</code>: specifies default genotyping error rates. </p><p><code>threshcall::Real = model == &quot;depmodel&quot; ? 0.95 : 0.9</code>: threshold for genotype calling. The filtering is based on called genotypes. </p><p><code>israndallele::Bool=true</code>: if true, genotyping error model follows the random allelic model, and otherwise the random genotypic model. </p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred, and otherwise they are outbred.</p><p><code>snpthin::Integer = 1</code>: take every snpthin-th markers.</p><p><code>ispermmarker::Bool=true</code>: if true, permute input marker ordering</p><p><code>isdupebinning::Union{Nothing,Bool}=false</code>: if ture, bin duplicate marker. </p><p><code>binshare::Real=0.5</code>: min fraction of shared genotypes between represent marker and each of the rest in a bin. </p><p><code>minlodsave::Union{Nothing, Real}=nothing</code>: results of pairwise analyses are saved only if the LOD score for LD or linkage &gt; minlodsave.    If it is nothing, minlodsave increases with number of markers. </p><p><code>minldsave::Union{Nothing, Real}=nothing</code>: results of pairwise LD analyses are saved only if the LD score (squared allelic correlation) &gt; minldsave.    If it is nothing, minldsave increases with number of markers. </p><p><code>ncluster::Union{Nothing, Integer}=nothing</code>: number of linkage groups.    If it is nothing, ncluster will be inferred in the range of [minncluster, maxncluster]</p><p><code>minncluster::Integer = isnothing(ncluster) ? 1 : ncluster</code>: min number of linkage groups.    If it is nothing, minncluster is set to 1 if ncluster = nothing and otherwise it is set to ncluster</p><p><code>maxncluster::Integer = isnothing(ncluster) ? 30 : ncluster</code>: max number of linkage groups.    If it is nothing, maxncluster is set to 30 if ncluster = nothing and otherwise it is set to ncluster</p><p><code>clusteralg::Union{Nothing,AbstractString}=nothing</code>: clustering algorithm after spectral embedding. </p><p><code>minsilhouette::Real=0.0</code>: delete markers withg silhouette scores &lt; minsilhouette. </p><p><code>minlodcluster::Union{Nothing, Real} = nothing</code>: minimum linkage LOD threshold.   If it is nothing, estimated internally as the minimum lod keeping the resulting   graph connected, ignoring the connected components of size &lt; mincomponentsize.</p><p><code>mincomponentsize::Union{Nothing,Integer} = nothing</code>: the markers in the graph   connectecd components of size &lt; mincomponentsize are removed. If it is nothing, it is internally set. </p><p><code>maxrf::Union{Nothing,Real} = nothing</code>: keep pairwise linakge analyses only if recombation fraction &lt;= maxrf. </p><p><code>binrf::Union{Nothing,Real}=nothing</code>: if binrf &gt;= 0, perform linkage-based marker binning such that the recombation fraction    for two markers in a bin is always &lt;= binrf, and otherwise not perform linkage-based binning.    If it is nothing, binrf is set to 0.001 if #markers &gt; ncluster<em>2000 and otherwise -1 if ncluster is not nothing,    and binrf is set to 0.001 if #markers &gt; maxncluster</em>2000 and otherwise -1 if ncluster is nothing. </p><p><code>alwayskeep::Real=0.99</code>: neighbors are always kept if its recombation fraction &gt;= alwayskeep, regardless of knncluster or knnorder. </p><p><code>minlodcluster::Union{Nothing,Real} = nothing</code>: min LOD score for clustering. If it is nothing, it is internally set. </p><p><code>minlodorder::Union{Nothing,Real} = nothing</code>: min LOD score for ordering. If it is nothing, it is internally set. </p><p><code>maxminlodcluster::Union{Nothing,Real} = 10,</code>: if minlodcluster = nothing, minlodcluster is internally estimated with upbound maxminlodcluster. </p><p><code>maxminlodorder::Union{Nothing,Real} = nothing</code>: if minlodorder = nothing, minlodorder is internally estimated with upbound maxminlodorder. </p><p><code>knncluster::Union{Nothing,Function} = nothing</code>: an anonymous function knncluster(x) of #markers x.    It returns #nearest neighbors for clustering. If it is nothing, knncluster = x-&gt;0.1*x. </p><p><code>knnorder::Union{Nothing,Function} = nothing</code>: an anonymous function knncluster(x) of #markers x in a linkage group.    It returns #nearest neighbors for ordering. If it is nothing, knnorder = x-&gt;sqrt(x). </p><p><code>isparallel::Bool=true</code>: if true, multicore computing over chromosomes.</p><p><code>workdir::AbstractString=pwd()</code>: working directory for input and output files.</p><p><code>commentstring::AbstractString=&quot;##&quot;</code>: rows that begin with commentstring will be ignored.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code>: stem of output filenames.</p><p><code>outext::AbstractString=&quot;.vcf.gz&quot;</code>: extension of output file for imputed geno.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= (isnothing(outstem) ? nothing : string(outstem,&quot;_magicimpute.log&quot;))</code>:   log file or IO for writing log. If it is nothing, no log file.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p><p><strong>Example</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicmap(genofile,pedinfo; ncluster=12)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicImpute" href="#MagicImpute"><code>MagicImpute</code></a> — <span class="docstring-category">Module</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicImpute</code></pre><p>a package for genotype imputation in multiparental populations.  Export two functions:  <a href="#MagicImpute.magicimpute"><code>magicimpute</code></a> and <a href="#MagicImpute.magicimpute!"><code>magicimpute!</code></a>.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicImpute.magicimpute" href="#MagicImpute.magicimpute"><code>MagicImpute.magicimpute</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicimpute(genofile, pedinfo;
    formatpriority, isphysmap, recomrate, commentstring,kwargs...)</code></pre><p>genotype imputation from genofile and pedinfo.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code> genotypic data file.</p><p><code>pedinfo::Union{MagicBase.JuncDist,AbstractString}</code> specifies pedigree information via a pedigree fille or a string designcode or via a struct juncdist::JuncDist.</p><p><strong>Keyword arguments</strong></p><p>See <a href="#MagicBase.formmagicgeno"><code>formmagicgeno</code></a> for the arguments (<code>formatpriority</code>, <code>isphysmap</code>, <code>recomrate</code>,<code>commentstring</code>) that are used for formming magicgeno.  Note that formatpriority=[&quot;AD&quot;,&quot;GT&quot;] by default. </p><p><code>mapfile::Union{Nothing, AbstractString}=nothing</code>: if it is nothing, use the marker map in the input genofile, and otherwise reset genetic marker map by that in mapfile.      The mapfile can either be in VCF format or in CSV format. For VCF format, genetic map is provided in the &quot;INFO&quot; column using keywords &quot;LINKAGEGROUP&quot; and &quot;POSCM&quot;.      For CSV-format, it must contain at least five columns: &quot;marker&quot;, &quot;linkagegroup&quot;, &quot;poscm&quot;, &quot;physchrom&quot;, and &quot;physposbp&quot;, where missing values are represented by &quot;NA&quot;.      If there exist columns &quot;binno&quot; and &quot;represent&quot;, markers with the same &quot;binno&quot; are binned with the represent being the marker with non-zero &quot;represent&quot;.      All the rest columns are ignored. </p><p>See <a href="#MagicImpute.magicimpute!"><code>magicimpute!</code></a> for the other arguments.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicimpute(&quot;geno.vcf.gz&quot;,&quot;4ril_self3&quot;)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicImpute.magicimpute!" href="#MagicImpute.magicimpute!"><code>MagicImpute.magicimpute!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicimpute!(magicgeno::MagicGeno; kwargs...)</code></pre><p>genotype imputation from magicgeno.</p><p><strong>Keyword arguments</strong></p><p><code>model::Union{AbstractString,AbstractVector}=&quot;jointmodel&quot;</code>:  prior depedence of ancestral prior process   along the two homologous chromosomes within an offspring. If model is a string, it must be &quot;depmodel&quot;,   &quot;indepmodel&quot;, or &quot;jointmodel&quot;. If model is a vector, the first element specifies the model    for founder imputation and the last element for offspring imputation. </p><p><code>likeparameters::LikeParameters=LikeParameters()</code>: parameters for genotypic data model.    If isinfererror = true, parameters with values being nothing will be inferred. </p><p><code>threshlikeparameters::ThreshLikeParameters=ThreshLikeParameters()</code>: markers with inferred likeparameters values &gt; threshlikeparameters values will be deleted. </p><p><code>priorlikeparameters::PriorLikeParameters=PriorLikeParameters()</code>: priors for likelihood parameters</p><p><code>israndallele::Bool=true</code>: if true, genotyping error model follows the random allelic model, and otherwise the random genotypic model. </p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred, and otherwise they are outbred.</p><p><code>chrsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of chromosome indices.   <code>nothing</code> denotes all chromosomes. Delete chromosome indices that are out of range.</p><p><code>snpsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of marker indices within each chromosome.   <code>nothing</code> denotes all markers. Marker indices that are larger than the number of markers   within the chromosome are deleted.</p><p><code>target::AbstractString = &quot;all&quot;</code>: target of imputation. <code>target=all</code> imputes   founders and offspring. <code>target</code> must be &quot;all&quot;, &quot;founder&quot;, &quot;offspring&quot;. </p><p><code>threshimpute::Real=0.9</code>: offspring genotypes are imputed if   the maximum posterior probability &gt; threshimpute.</p><p><code>byfounder::Integer=0</code>: alternatively impute each blocks of founders. The founders are partitioned such that the size of each block &lt;= byfounder.    If byfounder=-1, impute all founders simulteneously.    If byfounder=0, reset to the maximum subpopulation size, and the partition is based on the fouders of each sub-population.</p><p><code>inputneighbor::Union{Nothing,AbstractDict}=nothing</code>: nearest neighbors for each markers, which is required for neighbor-based marker order refinement.    If it is nothing and isordermarker = true, perform only random marker order refinement. </p><p><code>inputbinning::Union{Nothing,AbstractDict}=nothing</code>: a parition of markers into bins.  If it is not nothing, first impute founders for representative markers of each bin    and them impute founders for all markers. </p><p><code>isinfererror::Bool = true</code>: if true, infer marker specific likelihood parameters that have values of nothing in likeparameters. </p><p><code>tukeyfence::Real=3.0</code>: tukey fence for detecting outlier error rates (including foundererror, offspringerror, seqerror, and allelebalancemean). </p><p><code>minoutlier::Real=0.05</code>: markers with outlier error rates are removed only if their error rates &gt; minoutlier. </p><p><code>iscorrectfounder::Union{Nothing, Bool} = nothing</code>: if true, perform parental error correction.</p><p><code>phasealg::AbstractString=&quot;unphase&quot;</code>: if phasealg=forwardbackward, the output diplotype probabilities (in format GP), corresonding to the phased genotypes 0|0, 0|1, 1|0, and 1|1, are caculated based on the forward-backward algorithm,    and the output phased offspring genotypes (in format GT) are given by those with the largest diplotype probabilities if they are greater than threshcall.    If phasealg=viterbi, the output diplotype probabilities (GP) are set to those of phasealg=forwardbackward, and the output phased genotypes (GT) are caculated based on the Viterbi algorithm.    If phasealg=unphase, the output genotype probabilities (GP), corresonding to the unphased genotypes 0/0, 0/1, and 1/1, are calculated based on the forward backward algorithm,    and the output unphased genotypes (GT) are given by those with the largest genotype probabilities if they are greater than threshcall. </p><p><code>isdelmarker::Bool=true</code>: if true, perform marker deletion.</p><p><code>delsiglevel::Real = 0.01</code>: significance level for marker deletion</p><p><code>isordermarker::Bool=false</code>: if true, refine local marker ordering.</p><p><code>isspacemarker::Bool=false</code>: if true, estimate inter-marker distances.</p><p><code>trimcm::Real=20</code>: remove markers of each segment with distances to the flanking markers &gt; trimcm.   The number of markers of each segment must be less than 5% total number of markers.</p><p><code>skeletonsize::Union{Nothing,Integer} = nothing</code>: number of skeleton markers for piecewisely re-scaling inter-marker distances.    If it is nothing, skeletonsize is set to the number of distint positions in the genetic map before re-scaling. </p><p><code>slidewin_neighbor::Union{Nothing,Integer} = 200</code>: max sliding window size for neighbor-based marker order refinement. </p><p><code>slidewin::Union{Nothing,Integer} = nothing</code>: max sliding window size for random marker order refinement </p><p><code>binriffle::Union{Nothing,Integer} = nothing</code>: valid only in the case of marker binning. Skip magicimpute<em>founder after replacing representatives with binned markers if binriffle &lt; 0.    Keep magicimpute</em>founder for binned markers but without refinning ordering if 0 &lt;= binriffle &lt;= 1.    Keep magicimpute_founder for binned markers if binriffle &gt;=2, and if isordermarker = true set random order refinement with slidewin = binriffle and ignore neighbor-based order refinement. </p><p><code>orderactions::AbstractVector = [&quot;inverse&quot;,&quot;inverse00&quot;]</code>: update actions for random marker order refinement.    It must be a subset of [&quot;inverse&quot;,&quot;inverse00&quot;, &quot;inverse01&quot;,&quot;inverse10&quot;]. </p><p><code>orderactions_neighbor::AbstractVector = [&quot;inverse&quot;,&quot;inverse01&quot;]</code>: update actions for neighbor-based marker order refinement.    It must be a subset of [&quot;inverse&quot;,&quot;inverse00&quot;, &quot;inverse01&quot;,&quot;inverse10&quot;]. </p><p><code>inittemperature::Real= isordermarker ? 2.0 : 0.0</code>: initial temperature of annealing algorithm for marker ordering.</p><p><code>coolrate::Real=0.7</code>: temperature is mutiplied by coolrate after each iteration of annealing agrogrithm.</p><p><code>minaccept::Real=0.15</code>: minimum accept rate for controlling the window size of ordering update.</p><p><code>isparallel::Bool=true</code>: if true, parallel multicore computing over chromosomes.</p><p><code>workdir::AbstractString=pwd()</code>: working directory for input and output files.</p><p><code>tempdirectory::AbstractString = tempdir()</code>: temporary directory for inter-mediate results.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code>: stem of output filenames.</p><p><code>outext::AbstractString=&quot;.vcf.gz&quot;</code>: extension of output file for imputed geno.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= (isnothing(outstem) ? nothing : string(outstem,&quot;_magicimpute.log&quot;))</code>:   log file or IO for writing log. If it is nothing, no log file.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicgeno = formmagicgeno(&quot;geno.vcf.gz&quot;,&quot;ped.csv&quot;)
julia&gt; magicimpute!(magicgeno)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicReconstruct" href="#MagicReconstruct"><code>MagicReconstruct</code></a> — <span class="docstring-category">Module</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">MagicReconstruct</code></pre><p>a package for haplotype reconstruction in connected multiparental populations.  Export two functions:  <a href="#MagicReconstruct.magicreconstruct"><code>magicreconstruct</code></a> and <a href="#MagicReconstruct.magicreconstruct!"><code>magicreconstruct!</code></a>.</p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicReconstruct.magicreconstruct" href="#MagicReconstruct.magicreconstruct"><code>MagicReconstruct.magicreconstruct</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicreconstruct(genofile, pedinfo;
    formatpriority, isphysmap, recomrate, commentstring,kwargs...)</code></pre><p>haplotye reconstruction from genofile and pedinfo.</p><p><strong>Positional arguments</strong></p><p><code>genofile::AbstractString</code> genotypic data file.</p><p><code>pedinfo::Union{MagicBase.JuncDist,AbstractString}</code> specifies pedigree information via a pedigree fille or a string designcode or via a struct juncdist::JuncDist.</p><p><strong>Keyword arguments</strong></p><p>See <a href="#MagicBase.formmagicgeno"><code>formmagicgeno</code></a> for the arguments (<code>formatpriority</code>, <code>isphysmap</code>, <code>recomrate</code>,<code>commentstring</code>) that are used for formming magicgeno,  except that formatpriority=[&quot;GP&quot;, &quot;AD&quot;, &quot;GT&quot;] by default. </p><p>See <a href="#MagicReconstruct.magicreconstruct!"><code>magicreconstruct!</code></a> for the other arguments.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicreconstruct(&quot;geno.vcf.gz&quot;,&quot;4ril_self3&quot;)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicReconstruct.magicreconstruct!" href="#MagicReconstruct.magicreconstruct!"><code>MagicReconstruct.magicreconstruct!</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicreconstruct!(magicgeno::MagicGeno; kwargs...)</code></pre><p>haplotype reconstruction from magicgeno.</p><p><strong>Keyword arguments</strong></p><p><code>model::AbstractString=&quot;jointmodel&quot;</code>:  prior depedence of ancestral prior process   along the two homologous chromosomes within an offspring. It must be &quot;depmodel&quot;,   &quot;indepmodel&quot;, or &quot;jointmodel&quot;. </p><p><code>israndallele::Bool=true</code>: if true, genotyping error model follows the random allelic model, and otherwise the random genotypic model. </p><p><code>isfounderinbred::Bool=true</code>: if true, founders are inbred, and otherwise they are outbred.</p><p><code>chrsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of chromosome indices.   <code>nothing</code> denotes all chromosomes. Delete chromosome indices that are out of range.</p><p><code>snpsubset::Union{Nothing,AbstractRange,AbstractVector}=nothing</code>: subset of marker indices within each chromosome.   <code>nothing</code> denotes all markers. Marker indices that are larger than the number of markers   within the chromosome are deleted.</p><p><code>hmmalg::AbstractString=&quot;forwardbackward&quot;</code>: HMM alogrithm for haplotype   reconstruction, and it must be either &quot;forwardbackward&quot; or &quot;viterbi&quot;.</p><p><code>isignorephase::Bool=false</code>: if true, the phases of offspring genotypes are ignored. </p><p><code>isMMA::Bool=true</code>: if true, the Mathematica version of RABBIT.</p><p><code>nplot_subpop::Integer=10</code>: plots for up to nplot_subpop offspring in each subpopulation. </p><p><code>isparallel::Bool=true</code>: if true, multicore computing over chromosomes.</p><p><code>workdir::AbstractString=pwd()</code>: working directory for input and output files.</p><p><code>tempdirectory::AbstractString = tempdir()</code>: temporary directory for inter-mediate results.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code>: stem of output filenames.</p><p><code>outext::AbstractString=&quot;.csv.gz&quot;</code>: extension of output file for imputed geno.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= (isnothing(outstem) ? nothing : string(outstem,&quot;_magicimpute.log&quot;))</code>:   log file or IO for writing log. If it is nothing, no log file.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicgeno = formmagicgeno(&quot;geno.vcf.gz&quot;,&quot;ped.csv&quot;)
julia&gt; magicreconstruct(magicgeno,model=&quot;jointmodel&quot;)</code></pre></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicScan" href="#MagicScan"><code>MagicScan</code></a> — <span class="docstring-category">Module</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><p>MagicScan </p><p>a package for genomic QTL scan in connected multiparental populations.  Export one function: <a href="#MagicScan.magicscan"><code>magicscan</code></a>. </p></div></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicScan.magicscan" href="#MagicScan.magicscan"><code>MagicScan.magicscan</code></a> — <span class="docstring-category">Function</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">magicscan(ancestryfile, phenofile; commentstring=&quot;##&quot;,kwargs...)</code></pre><p>perform genomics scan of QTL for multiparental populations. See <code>magicreconstruct</code> for generating ancestryfile by haplotype reconstruction.</p><p><strong>Positional arguments</strong></p><p><code>ancestryfile::AbstractString</code> specifies ancestry file resulting <code>magicreconstruct</code></p><p><code>phenofile::AbstractString</code> specifies phenotypic file.</p><p><strong>Keyword arguments</strong></p><p><code>equation::Union{Nothing,FormulaTerm} = nothing</code> speficies linear model equation. By defulat, StatsModel.@formula(y ~ 1), where y is the last column name in phenofile.</p><p><code>thresholds::Union{Nothing,AbstractVector} = nothing</code>: list of thresholds used in plotting scanning profile.</p><p><code>nperm::Integer=200</code>: number of permutations of phenotypes  for calculating thresholds that are not specified.</p><p><code>siglevels::AbstractVector = [0.05]</code>: significance levels for calculating thresholds by permutations.</p><p><code>islog10p::Bool=true</code>: if true, the profile refers to -log10P, and LOD otherwise.</p><p><code>missingstring=[&quot;NA&quot;,&quot;missing&quot;]</code>: string denotes a missing phenotypic value.</p><p><code>commentstring::AbstractString</code> specifies the lines beginning with commentstring are ignored in genofile or pedfile given by <code>pedinfo</code>.</p><p><code>workdir::AbstractString=pwd()</code>: working directory for input and output files.</p><p><code>outstem::Union{Nothing,AbstractString}=&quot;outstem&quot;</code> specifies the stem of filename saving magicancestry. See <a href="#MagicBase.savemagicancestry"><code>MagicBase.savemagicancestry</code></a> for the description of outputfile &quot;outstem_magicancestry.csv.gz&quot;.</p><p><code>logfile::Union{Nothing,AbstractString,IO}= (isnothing(outstem) ? nothing : string(outstem,&quot;_magicreconstruct.log&quot;))</code>: log file or IO for writing log. If it is nothing, no log file.</p><p><code>verbose::Bool=true</code>: if true, print details on the stdout.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; magicscan(&quot;magicancestry.csv.gz&quot;,&quot;pheno.csv&quot;)</code></pre></div></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../rabbit_cli/">« CLI</a><a class="docs-footer-nextpage" href="../ref/">Citations »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Tuesday 7 January 2025 16:47">Tuesday 7 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
