<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Key options · RABBIT</title><meta name="title" content="Key options · RABBIT"/><meta property="og:title" content="Key options · RABBIT"/><meta property="twitter:title" content="Key options · RABBIT"/><meta name="description" content="Documentation for RABBIT."/><meta property="og:description" content="Documentation for RABBIT."/><meta property="twitter:description" content="Documentation for RABBIT."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../">RABBIT</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox" checked/><label class="tocitem" for="menuitem-2"><span class="docs-label">Guide</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../run_rabbit/">Run RABBIT</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Prepare Inputs</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../prepare_input/prepare_genofile/">Prepare genofile</a></li><li><a class="tocitem" href="../prepare_input/prepare_pedinfo_designcode/">Prepare pedinfo: designcode</a></li><li><a class="tocitem" href="../prepare_input/prepare_pedinfo_pedfile/">Prepare pedinfo: pedfile</a></li><li><a class="tocitem" href="../prepare_input/prepare_phenofile/">Prepare phenofile</a></li><li><a class="tocitem" href="../prepare_input/prepare_convert/">Prepare by converting </a></li></ul></li><li><a class="tocitem" href="../understand_output/">Understand outputs</a></li><li class="is-active"><a class="tocitem" href>Key options</a><ul class="internal"><li><a class="tocitem" href="#Genotype-formats"><span>Genotype formats</span></a></li><li><a class="tocitem" href="#Likelihood-parameters"><span>Likelihood parameters</span></a></li></ul></li><li><a class="tocitem" href="../statistical_framework/">Statistical framework</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox"/><label class="tocitem" for="menuitem-3"><span class="docs-label">Pipeline</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../pipeline/s0_pipeline_overview/">Overview</a></li><li><a class="tocitem" href="../pipeline/s1_magicsimulate/">S1 MagicSimulate</a></li><li><a class="tocitem" href="../pipeline/s2_magicfilter/">S2 MagicFilter</a></li><li><a class="tocitem" href="../pipeline/s3_magiccall/">S3 MagicCall</a></li><li><a class="tocitem" href="../pipeline/s4_magicmap/">S4 MagicMap</a></li><li><a class="tocitem" href="../pipeline/s5_magicimpute/">S5 MagicImpute</a></li><li><a class="tocitem" href="../pipeline/s6_magicreconstruct/">S6 MagicReconstruct</a></li><li><a class="tocitem" href="../pipeline/s7_magicscan/">S7 MagicScan</a></li></ul></li><li><a class="tocitem" href="../rabbit_cli/">CLI</a></li><li><a class="tocitem" href="../rabbit_api/">API</a></li><li><a class="tocitem" href="../ref/">Citations</a></li></ul><div class="docs-version-selector field has-addons visible"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"><option value="#" selected="selected">v1.0.0</option></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Guide</a></li><li class="is-active"><a href>Key options</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>Key options</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="Key-options"><a class="docs-heading-anchor" href="#Key-options">Key options</a><a id="Key-options-1"></a><a class="docs-heading-anchor-permalink" href="#Key-options" title="Permalink"></a></h1><h2 id="Genotype-formats"><a class="docs-heading-anchor" href="#Genotype-formats">Genotype formats</a><a id="Genotype-formats-1"></a><a class="docs-heading-anchor-permalink" href="#Genotype-formats" title="Permalink"></a></h2><p>The VCF genofile allows to specify multiple genotype formats for each marker. RABBIT assumes that all markers are biallelic; multiallelic markers can be either deleted (delmultiallelic=true) or transformed into biallelic markers by pooling multiple alternative alleles. RABBIT allows three kinds of genotype formats:</p><ul><li>GT: discrete genotype such as &quot;0/1&quot; for unphased genotypes and &quot;0|1&quot;  for phased genotype.</li><li>AD: allelic depth in form of &quot;r0,r1&quot;, where r0 is the number of reads for allele 0 and r1 is the number of reads for allele 1.</li><li>GP: genotype probability. It can be either in form of &quot;p00, p01, p11&quot;, where p00, p01, and p11 denote the probabilities of being genotype &quot;0/0&quot;, &quot;0/1&quot;, and &quot;1/1&quot;, respectively; or in form of &quot;p00, p11&quot; for inbred lines with only two possible homozygous genotypes (model=depmodel for <code>magicimpute</code>).</li></ul><p>RABBIT has a keyword argument <code>formatpriority</code> for functions requiring genofile as an argument. Specify <code>formatpriority</code> as a vector of formats with <strong>descreasing</strong>  priorities. To optimize the performance, the default value of <code>formatpriority</code> varies with function.</p><table><tr><th style="text-align: left">Function</th><th style="text-align: left">formatpriority</th></tr><tr><td style="text-align: left">magicfilter</td><td style="text-align: left">[&quot;AD&quot;,&quot;GT&quot;]</td></tr><tr><td style="text-align: left">magicmap</td><td style="text-align: left">[&quot;GT&quot;,&quot;AD&quot;]</td></tr><tr><td style="text-align: left">magicimpute</td><td style="text-align: left">[&quot;AD&quot;,&quot;GT&quot;]</td></tr><tr><td style="text-align: left">magicreconstruct</td><td style="text-align: left">[&quot;GP&quot;,&quot;AD&quot;,&quot;GT&quot;]</td></tr></table><div class="admonition is-info"><header class="admonition-header">formatpriority</header><div class="admonition-body"><ul><li>Specify formatpriority like this <code>--formatpriority &quot;[AD, GT]&quot;</code> for running in a command, and <code>formatpriority = [&quot;AD&quot;, &quot;GT&quot;]</code> for running in julia REPL.</li><li><code>magiccall</code> performs single marker genotype calling. For each marker with AD, it first infers parental genotypes and error rates, and then exports AD, calculated GP, and called GT.</li><li><code>magicmap</code> performs raw genotype calling for &quot;AD&quot; if skipping <code>magiccall</code> for sequence data. The raw genotype calling assumes a pre-specified error rate, and it performs well for almost homozygous populations, but not for sequence data from heterozygous populations.</li><li><code>magicimpute</code> exports both posterior genotype probabilities (&quot;GP&quot;) and discrete genotypes (&quot;GT&quot;).</li><li><code>magicreconstruct</code> performs generally better if using &quot;GP&quot; than that of &quot;GT&quot;.</li></ul></div></div><h2 id="Likelihood-parameters"><a class="docs-heading-anchor" href="#Likelihood-parameters">Likelihood parameters</a><a id="Likelihood-parameters-1"></a><a class="docs-heading-anchor-permalink" href="#Likelihood-parameters" title="Permalink"></a></h2><p>Genotype likelihood denotes the probability of observed genotypic data given hidden states and parameters.  The parameters include <code>foundererror</code>, <code>offspringerror</code>, <code>seqerror</code>, <code>allelebalancemean</code>, <code>allelebalancedisperse</code>, and <code>alleledropout</code>. See the section <a href="../statistical_framework/#statistical-framework"><code>Statistical Framework</code></a>.</p><p>RABBIT has three keyword arguments <code>likeparameters</code>, <code>priorlikeparameters</code>, and <code>threshlikeparameters</code>, which have struct types <code>LikeParameters</code>, <code>PriorLikeParameters</code>, and <code>TreshLikeParameters</code>, respectively, and each of the structs has field names for the six likelihood parameters. </p><h3 id="LikeParameters"><a class="docs-heading-anchor" href="#LikeParameters">LikeParameters</a><a id="LikeParameters-1"></a><a class="docs-heading-anchor-permalink" href="#LikeParameters" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.LikeParameters" href="#MagicBase.LikeParameters"><code>MagicBase.LikeParameters</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">LikeParameters</code></pre><p>keyword-based struct for the parameters of likelihood function. </p><p>LikeParameters() is equivalent to LikeParameters(foundererror=0.005, offspringerror=nothing, peroffspringerror=nothing,  seqerror=nothing, allelebalancemean=nothing, allelebalancedisperse=nothing, alleledropout=0.0).  The <code>peroffspringerror</code> refers to error rate per offspring, and the other parameters refer to error rate per marker. </p><p>If genotype format is not &quot;AD&quot;, the parameters <code>seqerror</code>, <code>allelebalancemean</code>, <code>allelebalancedisperse</code>, and <code>alleledropout</code> are irrelevant. </p><p>If <code>model=&quot;depmodel&quot;</code>, the parameters <code>allelebalancemean</code>, <code>allelebalancedisperse</code>, and <code>alleledropout</code> are irrelevant. </p><p>If there exists keyarg <code>isinfererror</code> and <code>isinfererror</code> = true, the parameters  with values being nothing will be inferred and the other parameters will be fixed.   If <code>isinfererror</code> = false,  LikeParameters() is equivalent to LikeParameters(foundererror=0.005, offspringerror=0.005, peroffspringerror=0.0,  seqerror=0.001, allelebalancemean=0.5, allelebalancedisperse=0.0, alleledropout=0.0). </p></div></section></article><h3 id="PriorLikeParameters"><a class="docs-heading-anchor" href="#PriorLikeParameters">PriorLikeParameters</a><a id="PriorLikeParameters-1"></a><a class="docs-heading-anchor-permalink" href="#PriorLikeParameters" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.PriorLikeParameters" href="#MagicBase.PriorLikeParameters"><code>MagicBase.PriorLikeParameters</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">PriorLikeParameters</code></pre><p>keyword-based struct for the priors of the likelihood parameters. </p><p><strong>Fields</strong></p><p><code>foundererror::Distribution = Beta(1.0,1.0)</code>: prior distribution of founder error rate</p><p><code>offspringerror::Distribution = Beta(1.0,1.0)</code>: prior distribution of offspring error rate</p><p><code>peroffspringerror::Distribution = Beta(1.0,1.0)</code>: prior distribution of error rate per offspring </p><p><code>seqerror::Distribution = Beta(1.0,1.0)</code>: prior distribution of sequence base error rate</p><p><code>allelebalancemean::Distribution = Beta(1.01,1.01)</code> prior distribution of allele balance mean</p><p><code>allelebalancedisperse::Distribution = Exponential(0.5)</code>: prior distribution of allele balance overdispersion</p><p><code>alleledropout::Distribution = Beta(1,19)</code>: prior distribution of allele dropout rate</p></div></section></article><h3 id="ThreshLikeParameters"><a class="docs-heading-anchor" href="#ThreshLikeParameters">ThreshLikeParameters</a><a id="ThreshLikeParameters-1"></a><a class="docs-heading-anchor-permalink" href="#ThreshLikeParameters" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="MagicBase.ThreshLikeParameters" href="#MagicBase.ThreshLikeParameters"><code>MagicBase.ThreshLikeParameters</code></a> — <span class="docstring-category">Type</span><span class="is-flex-grow-1 docstring-article-toggle-button" title="Collapse docstring"></span></header><section><div><pre><code class="language-julia hljs">ThreshLikeParameters</code></pre><p>keyword-based struct for the thresholds of the likelihood parameters. Markers with the inferred parameter values  being greater than the maximum will be deleted. </p><p>ThreshLikeParameters() is equivalent to ThreshLikeParameters(foundererror=0.25, offspringerror=0.25,  peroffspringerror=0.25, seqerror=0.25, allelebalancemean=0.9, allelebalancedisperse=1.0, alleledropout=0.05). </p></div></section></article><div class="admonition is-info"><header class="admonition-header">Likelihood parameters</header><div class="admonition-body"><ul><li><code>magicimpute</code> and <code>magiccall</code> have the three keyargs for likelihood parameters and keyarg <code>isinfererror</code>.</li><li><code>magicimpute</code> has default <code>likeparameters = LikeParameters()</code>, <code>priorlikeparameters=PriorLikeParameters()</code>, and <code>threshlikeparameters = ThreshLikeParameters()</code></li><li><code>magiccall</code> has default <code>likeparameters = LikeParameters(offspringerror=0.04, alleledropout=0.0)</code>, <code>priorlikeparameters=PriorLikeParameters(seqerror=0.005)</code>, and <code>threshlikeparameters = ThreshLikeParameters(seqerror=0.05)</code></li></ul></div></div></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../understand_output/">« Understand outputs</a><a class="docs-footer-nextpage" href="../statistical_framework/">Statistical framework »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Tuesday 7 January 2025 16:05">Tuesday 7 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
