<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>S1 MagicSimulate · RABBIT</title><meta name="title" content="S1 MagicSimulate · RABBIT"/><meta property="og:title" content="S1 MagicSimulate · RABBIT"/><meta property="twitter:title" content="S1 MagicSimulate · RABBIT"/><meta name="description" content="Documentation for RABBIT."/><meta property="og:description" content="Documentation for RABBIT."/><meta property="twitter:description" content="Documentation for RABBIT."/><script data-outdated-warner src="../../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../search_index.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-mocha.css" data-theme-name="catppuccin-mocha"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-macchiato.css" data-theme-name="catppuccin-macchiato"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-frappe.css" data-theme-name="catppuccin-frappe"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/catppuccin-latte.css" data-theme-name="catppuccin-latte"/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><div class="docs-package-name"><span class="docs-autofit"><a href="../../">RABBIT</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><a class="tocitem" href="../../">Home</a></li><li><input class="collapse-toggle" id="menuitem-2" type="checkbox"/><label class="tocitem" for="menuitem-2"><span class="docs-label">Guide</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../run_rabbit/">Run RABBIT</a></li><li><input class="collapse-toggle" id="menuitem-2-2" type="checkbox"/><label class="tocitem" for="menuitem-2-2"><span class="docs-label">Prepare Inputs</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../../prepare_input/prepare_genofile/">Prepare genofile</a></li><li><a class="tocitem" href="../../prepare_input/prepare_pedinfo_designcode/">Prepare pedinfo: designcode</a></li><li><a class="tocitem" href="../../prepare_input/prepare_pedinfo_pedfile/">Prepare pedinfo: pedfile</a></li><li><a class="tocitem" href="../../prepare_input/prepare_phenofile/">Prepare phenofile</a></li><li><a class="tocitem" href="../../prepare_input/prepare_convert/">Prepare by converting </a></li></ul></li><li><a class="tocitem" href="../../understand_output/">Understand outputs</a></li><li><a class="tocitem" href="../../key_options/">Key options</a></li><li><a class="tocitem" href="../../statistical_framework/">Statistical framework</a></li></ul></li><li><input class="collapse-toggle" id="menuitem-3" type="checkbox" checked/><label class="tocitem" for="menuitem-3"><span class="docs-label">Pipeline</span><i class="docs-chevron"></i></label><ul class="collapsed"><li><a class="tocitem" href="../s0_pipeline_overview/">Overview</a></li><li class="is-active"><a class="tocitem" href>S1 MagicSimulate</a></li><li><a class="tocitem" href="../s2_magicfilter/">S2 MagicFilter</a></li><li><a class="tocitem" href="../s3_magiccall/">S3 MagicCall</a></li><li><a class="tocitem" href="../s4_magicmap/">S4 MagicMap</a></li><li><a class="tocitem" href="../s5_magicimpute/">S5 MagicImpute</a></li><li><a class="tocitem" href="../s6_magicreconstruct/">S6 MagicReconstruct</a></li><li><a class="tocitem" href="../s7_magicscan/">S7 MagicScan</a></li></ul></li><li><a class="tocitem" href="../../rabbit_cli/">CLI</a></li><li><a class="tocitem" href="../../rabbit_api/">API</a></li><li><a class="tocitem" href="../../ref/">Citations</a></li></ul><div class="docs-version-selector field has-addons visible"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"><option value="#" selected="selected">v1.0.0</option></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li><a class="is-disabled">Pipeline</a></li><li class="is-active"><a href>S1 MagicSimulate</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>S1 MagicSimulate</a></li></ul></nav><div class="docs-right"><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="S1-MagicSimulate"><a class="docs-heading-anchor" href="#S1-MagicSimulate">S1 MagicSimulate</a><a id="S1-MagicSimulate-1"></a><a class="docs-heading-anchor-permalink" href="#S1-MagicSimulate" title="Permalink"></a></h1><h3 id="Data-Simulation"><a class="docs-heading-anchor" href="#Data-Simulation">Data Simulation</a><a id="Data-Simulation-1"></a><a class="docs-heading-anchor-permalink" href="#Data-Simulation" title="Permalink"></a></h3><p>Set the working directory to the directory of this file. All functions that have keyarg outstem (or outfile) have keyarg <code>workdir</code>, and by default <code>workdir</code> is the current directory (<code>pwd()</code>).</p><pre><code class="language-julia hljs"># code for Julia
cd(@__DIR__)  </code></pre><p>For using RABBIT in CLI, use command <code>cd WORKDIR</code> for changing directory to the <code>WORKDIR</code> containing input files. </p><h3 id="Simulate-founder-data"><a class="docs-heading-anchor" href="#Simulate-founder-data">Simulate founder data</a><a id="Simulate-founder-data-1"></a><a class="docs-heading-anchor-permalink" href="#Simulate-founder-data" title="Permalink"></a></h3><p>Simulate genotypic data for 4 inbred founders with 400 markers randomly distributed over 4 chromosomes; each chromosome has length 100 cM. </p><pre><code class="language-julia hljs"># code for Julia
using MagicSimulate
outstem = &quot;example&quot;
fhaplofile = outstem*&quot;_fhaplo.vcf.gz&quot;    
simfhaplo(;
    isfounderinbred = true, # default
    nsnp=400, 
    nparent=4,
    chrlen=100*ones(4),
    outfile=fhaplofile
)</code></pre><p>For running RABBIT in CLI, use command <code>cd WORKDIR</code> for change directory to the <code>WORKDIR</code> containing input files. </p><pre><code class="language-bash hljs"># code for Linux shell. 
# For Window CMD, replace multiline key \ by  ^, and replace comment-key # by ::
julia rabbit_simfhaplo.jl --isfounderinbred  true --nparent 4 \
    --chrlen &quot;[100,100,100,100]&quot; \
    --nsnp 300 \    
    -o example_fhaplo.vcf.gz</code></pre><h3 id="Simulate-pedigree-information"><a class="docs-heading-anchor" href="#Simulate-pedigree-information">Simulate pedigree information</a><a id="Simulate-pedigree-information-1"></a><a class="docs-heading-anchor-permalink" href="#Simulate-pedigree-information" title="Permalink"></a></h3><p>Simulate a breeding pedigree: 3 subpopulations (DH, F2, and three-parent cross) with equal size 100. </p><pre><code class="language-julia hljs"># code for Julia
using MagicBase
magicped = generate_magicped(;
  designcodes=[&quot;P1/P2=&gt;DH&quot;, &quot;2ril-self1&quot;, &quot;P4/3/P4//P2/P3=&gt;1&quot;],
  founders = [&quot;NA&quot;,&quot;P1||P3&quot;,&quot;NA&quot;],
  subpopsizes=100*ones(3)
)
savemagicped(outstem*&quot;_ped.csv&quot;, magicped)</code></pre><pre><code class="language-bash hljs"># code for Linux shell. 
# For Window CMD, replace multiline key \ by  ^, and replace comment-key # by ::
julia rabbit_generate_magicped.jl --designcodes  &quot;[P1/P2=&gt;DH, 2ril-self1, P4/3/P4//P2/P3=&gt;1]&quot;  \
    --founders &quot;[NA, P1||P3, NA]&quot; \
    --subpopsizes &quot;[100,100,100]&quot; \
    -o example</code></pre><h3 id="Simulate-offspring-data"><a class="docs-heading-anchor" href="#Simulate-offspring-data">Simulate offspring data</a><a id="Simulate-offspring-data-1"></a><a class="docs-heading-anchor-permalink" href="#Simulate-offspring-data" title="Permalink"></a></h3><p>Simulate genotypic data for the example connected multiparental population. If ispheno=true, simulate phenotypic values with a specified heritability and the number of QTLs underlying the trait.</p><pre><code class="language-julia hljs"># code for Julia
using MagicSimulate
using Distributions
pedfile = outstem*&quot;_ped.csv&quot;
magicsimulate(fhaplofile,pedfile;    
    seqfrac = 1.0,
    seqdepth = Gamma(2,5),
    foundermiss = Beta(1,9),
    offspringmiss = Beta(1,9),
    foundererror = Beta(1,19),
    offspringerror = Beta(1,19),    
    allelebalancemean = Beta(5,5),
    allelebalancedisperse = Exponential(0.05),    
    ispheno = true,
    pheno_nqtl=1,
    pheno_h2= 0.5,
    isplot = true, 
    outstem
)</code></pre><pre><code class="language-bash hljs"># code for Linux shell. 
# For Window CMD, replace multiline key \ by  ^, and replace comment-key # by ::
julia rabbit_magicsimulate.jl -g example_fhaplo.vcf.gz \
    -p example_ped.csv \
    --seqfrac 1.0 \
    --seqdepth Gamma(2,5) \
    --foundermiss Beta(1,9) \
    --offspringmiss Beta(1,9) \
    --foundererror Beta(1,19), \
    --offspringerror Beta(1,19) \
    --allelebalancemean Beta(5,5)  \
    --allelebalancedisperse Exponential(0.05) \    
    --ispheno true \
    --pheno_nqtl 1 \
    --pheno_h2 0.5 \
    -o example</code></pre><h3 id="Output-files"><a class="docs-heading-anchor" href="#Output-files">Output files</a><a id="Output-files-1"></a><a class="docs-heading-anchor-permalink" href="#Output-files" title="Permalink"></a></h3><table><tr><th style="text-align: left">outfile</th><th style="text-align: left">Description</th></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate.log&quot;</td><td style="text-align: left">log file</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_pheno.csv&quot;</td><td style="text-align: left">simulated observed phenofile</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_geno.vcf.gz&quot;</td><td style="text-align: left">simulated observed genofile</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_ped.csv&quot;</td><td style="text-align: left">simulated pedfile</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_truecontfgl.csv.gz&quot;</td><td style="text-align: left">simulated true inheritance pattern in continuous form</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_truecontfgl_mosaic.png&quot;</td><td style="text-align: left">plot genomic mosaic pattern (isplot=true)</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_truecontfgl_recom.png&quot;</td><td style="text-align: left">plot distribution of #breakpoints (isplot=true)</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_truefgl.csv.gz&quot;</td><td style="text-align: left">simulated true inheritance pattern at each marker</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_truegeno.csv.gz&quot;</td><td style="text-align: left">simulated true genofile</td></tr><tr><td style="text-align: left">outstem*&quot;_magicsimulate_truepheno.csv&quot;</td><td style="text-align: left">simulated true phenotype values</td></tr></table><h3 id="Output:-pedigree"><a class="docs-heading-anchor" href="#Output:-pedigree">Output: pedigree</a><a id="Output:-pedigree-1"></a><a class="docs-heading-anchor-permalink" href="#Output:-pedigree" title="Permalink"></a></h3><p><code>outstem*&quot;_magicsimulate_ped.csv&quot;</code> differs from the input <code>pedfile</code> by transforming the <code>designcodes</code> into pedigree format.  <code>outstem*&quot;_magicsimulate_ped.png&quot;</code> gives the plot of pedigree. </p><p><img src="../example_magicsimulate_ped.png" alt/></p><h3 id="Output:-mosaic"><a class="docs-heading-anchor" href="#Output:-mosaic">Output: mosaic</a><a id="Output:-mosaic-1"></a><a class="docs-heading-anchor-permalink" href="#Output:-mosaic" title="Permalink"></a></h3><p><code>outstem*&quot;_magicsimulate_truecontfgl_mosaic.png&quot;</code> plots genomic mosaic pattern for up to 3 offspring per subpopulation. </p><p><img src="../example_magicsimulate_truecontfgl_mosaic.png" alt/></p><h3 id="Output:-#breakpoints"><a class="docs-heading-anchor" href="#Output:-#breakpoints">Output: #breakpoints</a><a id="Output:-#breakpoints-1"></a><a class="docs-heading-anchor-permalink" href="#Output:-#breakpoints" title="Permalink"></a></h3><p><code>outstem*&quot;_magicsimulate_truecontfgl_recom.png&quot;</code> plots distribution of #recombination breakpoints. </p><p><img src="../example_magicsimulate_truecontfgl_recom.png" alt/></p></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../s0_pipeline_overview/">« Overview</a><a class="docs-footer-nextpage" href="../s2_magicfilter/">S2 MagicFilter »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option><option value="catppuccin-latte">catppuccin-latte</option><option value="catppuccin-frappe">catppuccin-frappe</option><option value="catppuccin-macchiato">catppuccin-macchiato</option><option value="catppuccin-mocha">catppuccin-mocha</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.8.0 on <span class="colophon-date" title="Tuesday 7 January 2025 16:05">Tuesday 7 January 2025</span>. Using Julia version 1.11.2.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
